{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c05e7e1a-9bd1-4dfc-93a9-b835d67f1a06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make tokens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a5861f1a-fb86-4911-a534-d99b9b8b39f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5d36e35a-ee93-420d-ab7b-6bcf3be63cc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in ['/opt/homebrew/Cellar/python@3.9/3.9.15/Frameworks/Python.framework/Versions/3.9/lib/python39.zip', \n",
    " '/opt/homebrew/Cellar/python@3.9/3.9.15/Frameworks/Python.framework/Versions/3.9/lib/python3.9', \n",
    " '/opt/homebrew/Cellar/python@3.9/3.9.15/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload', \n",
    " '/Users/christianbromley/.virtualenvs/machine-learning-course/lib/python3.9/site-packages']:\n",
    "    sys.path.insert(0, i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "157571c7-d139-4af9-9573-8b5e2b33a397",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/christianbromley/.virtualenvs/machine-learning-course/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import argparse\n",
    "from pathlib import Path\n",
    "\n",
    "#sys.path.insert(0,'./.env/lib/python3.8/site-packages')\n",
    "from tokenizers import ByteLevelBPETokenizer\n",
    "#from transformers import RobertaTokenizer\n",
    "import os\n",
    "\n",
    "from transformers import RobertaTokenizer, RobertaConfig, RobertaForMaskedLM, AdamW\n",
    "import torch\n",
    "import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "60617f40-ca82-4ef7-8418-fe3521871445",
   "metadata": {},
   "outputs": [],
   "source": [
    "# constants\n",
    "PDF_PATH='/Users/christianbromley/Documents/Personal/PhD/Christian_Bromley_Final_Thesis_20211221_cover.pdf'\n",
    "OUTPUT_PATH='/Users/christianbromley/Documents/Personal/PhD/Christian_Bromley_Final_Thesis_20211221.csv'\n",
    "TEXT_PARSE_DIR='/Users/christianbromley/Documents/Personal/PhD/thesis_parsed'\n",
    "MODELS_DIR='/Users/christianbromley/Documents/Projects/machine-learning-course/bert-transformer/models'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "daae5e06-ccfe-4a29-a910-9352c9b6792c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_file_paths(input_dir):\n",
    "    paths = [str(x) for x in Path(input_dir).glob('**/*.txt')]\n",
    "    return paths\n",
    "\n",
    "def run(paths, output_dir):\n",
    "    # intialise tokeniser\n",
    "    tokenizer = ByteLevelBPETokenizer()\n",
    "    # train\n",
    "    tokenizer.train(files=paths,\n",
    "                    vocab_size=30_522,\n",
    "                    min_frequency=2,\n",
    "                    special_tokens=['<s>', '<pad>', '</s>', '<unk>', '<mask>'])\n",
    "\n",
    "    tokenizer.save_model(output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "77d24587-e32f-414e-be1c-51e254ed29a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Users/christianbromley/Documents/Personal/PhD/thesis_parsed/text_0.txt']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paths = extract_file_paths(input_dir=TEXT_PARSE_DIR)\n",
    "\n",
    "paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ff5414c6-007a-42bc-b40d-9abab80b7f21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "run(paths, MODELS_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "72f7a40c-d257-4d8b-a7df-e3131cb28582",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The tokenizer class you load from this checkpoint is not the same type as the class this function is called from. It may result in unexpected tokenization. \n",
      "The tokenizer class you load from this checkpoint is 'BertTokenizer'. \n",
      "The class this function is called from is 'RobertaTokenizer'.\n"
     ]
    }
   ],
   "source": [
    "# initialize the tokenizer using the tokenizer we created\n",
    "tokenizer = RobertaTokenizer.from_pretrained(MODELS_DIR, max_len=512)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dd5b5d2c-9e13-4802-a3de-86478349b38e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def masked_language_model(tensor):\n",
    "    rand = torch.rand(tensor.shape)\n",
    "    mask_arr = (rand < 0.15) * (tensor > 2)\n",
    "    for i in range(tensor.shape[0]):\n",
    "        selection = torch.flatten(mask_arr[i].nonzero()).tolist()\n",
    "        tensor[i,selection] = 4\n",
    "    return tensor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "61c5f020-235e-41c0-ab96-89b72b508373",
   "metadata": {},
   "outputs": [],
   "source": [
    "## input ids - token IDs with a % of tokens masked with the mask token ID which in our case is 4\n",
    "input_ids = []\n",
    "## mask - this is a binary tensor of 1 and 0 indicating where the masks are\n",
    "mask = []\n",
    "## labels - these are just the unmasked token IDs\n",
    "labels = []\n",
    "# initialise the output\n",
    "tokenized_text = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9919ea8b-912f-4438-ad07-66cf54885c8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read file\n",
    "with open(paths[0], 'r', encoding='utf-8') as fp:\n",
    "    lines = fp.read().split('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "30c688f3-65c5-469e-8c75-a2dbc8019775",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['13Abstract  Cancer-related inflammation has classically been linked with tumour progression and therapy resistance.',\n",
       " 'T cell infiltration, a hallmark of so-called ‘hot’ inflamed tumours, is however associated with better prognosis and response to immune checkpoint blockade (ICB).',\n",
       " 'As such, inflammation can paradoxically promote and restrain tumour growth.',\n",
       " 'Yet, whether and how cancer-promoting inflammation impacts T cell-mediated tumour immunity is unclear.',\n",
       " 'Understanding the interplay of different inflammatory phenotypes is therefore a priority for the field.',\n",
       " 'The adoption of next generation sequencing technologies has fuelled a rapid increase in the amount of publicly available data from cancer and healthy tissue samples as well as pre-clinical models.',\n",
       " 'So far, pan-cancer meta-analyses using immunogenomic and transcriptomic data have focussed primarily on elucidating the characteristics of T cell-inflamed tumours, whilst systematic analyses of pro-tumourigenic inflammatory processes have been lacking.',\n",
       " 'The cyclooxygenase-2 (COX-2) pathway is frequently upregulated in cancer and has been linked to tumour-promoting inflammation.',\n",
       " 'Utilising this pathway as a central putative node of cancer-promoting inflammation, we derive a gene signature of extracellular inflammatory ligands in order to study the effect of pro-tumourigenic inflammation (PTI) on T cell-mediated immunity and cancer patient outcome.',\n",
       " 'By integrating multi-modal datasets from thousands of cancer patients and cell lines including in-house validation we hereby perform an extensive characterisation of intratumoural inflammatory phenotypes.',\n",
       " 'We show that expression of PTI genes correlates with oncogenic transformation, disease progression, resistance to ICB and has independent prognostic power across pan-cancer immune subtypes and individual malignancies.',\n",
       " 'Cancer cells themselves co-express PTI genes in association with KRAS mutations, MAPK and NFkB activation.',\n",
       " 'Accordingly, inhibition of MEK robustly attenuates secretion of PTI factors by cancer cells.',\n",
       " 'Analysis of patient single cell RNA sequencing datasets reveals that besides cancer cells, diverse stromal and leukocyte populations contribute towards PTI gene expression in tumours.',\n",
       " 'PTI associates with spatially distinct lymphocyte infiltration patterns and with T cell dysfunctional phenotypes and sub-optimal activation.',\n",
       " 'The correlation of T cell inflammation and IFNg signalling with patient survival is antagonised by PTI supporting a role for immune evasion in its profound prognostic power.',\n",
       " 'Collating transcriptomic data from patients treated with ICB, we find that PTI significantly associates with lack of response to ICB and exhibits joint predictive utility with PD-L1 and biomarkers of T cell inflammation.',\n",
       " 'In this analysis, PTI outperforms other signatures of ICB resistance.',\n",
       " 'Notably, patients with a high likelihood of responding to ICB according to current immune biomarkers can be further stratified by pre-treatment PTI levels.',\n",
       " 'Pan-cancer analysis shows that PTI has a major role in disease progression of clear cell renal cell carcinomas (ccRCC); a cancer type in which current biomarkers do not predict ICB response.',\n",
       " 'Through retrospective transcriptional profiling of archival ccRCC samples, and analysis of two independent validation cohorts, we show that the 16-gene PTI signature has potential to guide immmunotherapeutic decisions in ccRCC.',\n",
       " 'In doing so, we establish an experimental and bioinformatic pipeline for future translational work.',\n",
       " 'Collectively, our findings establish PTI as a critical determinant of tumour fate and response to immunotherapy beyond current immune biomarkers.  ',\n",
       " '14Declaration  No portion of the work referred to in this thesis has been submitted in support of an application for another degree or qualification of this or any other university or other institute of learning.',\n",
       " 'I.',\n",
       " 'In vitro cultivation of murine bone-marrow derived macrophages (BMDMs) and treatment with synthetic prostaglandin-E2 followed by RNA sequencing was carried out by Dr. Eimear Flanagan and Dr. Eduardo Bonavita in the Cancer Inflammation and Immunity laboratory at Cancer Research UK Manchester Institute. ',\n",
       " 'Copyright Statement  I.',\n",
       " 'The author of this thesis (including any appendices and/or schedules to this thesis) owns certain copyright or related rights in it (the “Copyright”) and s/he has given The University of Manchester certain rights to use such Copyright, including for administrative purposes.',\n",
       " 'II.',\n",
       " 'Copies of this thesis, either in full or in extracts and whether in hard or electronic copy, may be made only in accordance with the Copyright, Designs and Patents Act 1988 (as amended) and regulations issued under it or, where appropriate, in accordance with licensing agreements which the University has from time to time.',\n",
       " 'This page must form part of any such copies made.',\n",
       " 'III.',\n",
       " 'The ownership of certain Copyright, patents, designs, trademarks and other intellectual property (the “Intellectual Property”) and any reproductions of copyright works in the thesis, for example graphs and tables (“Reproductions”), which may be described in this thesis, may not be owned by the author and may be owned by third parties.',\n",
       " 'Such Intellectual Property and Reproductions cannot and must not be made available for use without the prior written permission of the owner(s) of the relevant Intellectual Property and/or Reproductions.',\n",
       " 'IV.',\n",
       " 'Further information on the conditions under which disclosure, publication and commercialisation of this thesis, the Copyright and any Intellectual Property and/or Reproductions described in it may take place is available in the University IP Policy (see http://documents.manchester.ac.uk/DocuInfo.aspx?DocID=2442 0), in any relevant Thesis restriction declarations deposited in the University Library, The University Library’s regulations (see http://www.library.manchester.ac.uk/about/regulations/) and in The University’s policy on Presentation of Theses.     ',\n",
       " '15Preface  The author has an undergraduate degree in Biomedical Science with Employment Experience from the University of Sheffield.',\n",
       " 'In addition to undergraduate laboratory and research experience, the author gained additional experience in industry spending one year working for GlaxoSmithKline and two years working for AstraZeneca predominantly in laboratory-based roles.',\n",
       " 'COVID-19 Impact Statement  This research project was impacted by the COVID-19 pandemic.',\n",
       " 'The Cancer Research UK Manchester Institute was closed between 16th March and 8th June 2020, and there was a time lag for getting the laboratory and core facility services running following the recommencement of work.',\n",
       " 'As such, there were severe delays in the experimental project work, whilst only in silico analyses could take place.',\n",
       " 'In addition, the work described herein to perform retrospective analysis of cancer patient samples requires a collaborative team of pathologists, clinicians, technicians and biobank staff.',\n",
       " 'This network was disrupted by the pandemic leading to delays in collecting and processing clinical samples as well as in obtaining critical clinical data.',\n",
       " 'The author was subsequently granted a three-month extension to the PhD project.          ',\n",
       " '16Introduction  Inflammation Immune cells collaborate to combat infection and repair damaged tissues.',\n",
       " 'The process by which immune cells infiltrate an infected or damaged site to elicit a return to homeostasis is called inflammation (Medzhitov, 2008).',\n",
       " 'Inflammation, therefore, is a protective mechanism by which the body defends itself from harmful stimuli.',\n",
       " 'A host of environmental cues can trigger an acute inflammatory response including pathogenic stimuli, commensals, cell death, hypoxia, foreign bodies, carcinogens, pollutants and radiation (Medzhitov, 2008).',\n",
       " 'These cues activate the innate immune system that continuously surveils different tissues.',\n",
       " 'Once activated, a rapid and coordinated process triggers the local production of cytokines, chemokines, prostaglandins and growth factors that act on local parenchymal cells, resident leukocytes, epithelial cells, and endothelial cells priming the tissue for an influx of immune cells (Nathan, 2002).',\n",
       " 'Activated endothelial cells control extravasation of neutrophils and monocytes into the tissue from the local capillaries (Pober and Sessa, 2007).',\n",
       " 'Simultaneously, secreted signalling molecules diffuse into the circulation to elicit systemic inflammatory changes such as stimulating the production of new bone marrow-derived immune cells in order to fuel the local inflammatory cascade.',\n",
       " 'In physiological settings, this cascade of events culminates in the destruction of infected cells or pathogens, followed by a careful reconstruction and repair of the tissue, damaged in the first instance by an external stimulus and secondly through the collateral damage inflicted by the inflammatory response.',\n",
       " 'This coordinated effort to return the tissue to homeostasis is frequently called the wound healing response or the resolution phase of inflammation (Serhan and Savill, 2005).   ',\n",
       " 'Failure to resolve acute inflammation leads to chronic inflammation that has a unique set of characteristics.',\n",
       " 'This chronic inflammatory state shifts the homeostatic threshold leading to the manifestation of diverse pathologies (Nathan, 2002).',\n",
       " 'The nature of this state depends on the initial inducer of inflammation, whether pathogenic or endogenous, microbial or non-microbial, and the mechanisms deployed to resolve the injury during the acute phase response.',\n",
       " 'For example, autoimmune conditions like rheumatoid arthritis are the result of a sustained inflammatory response against self-antigens (Horai et al., 2000).',\n",
       " 'Moreover, in certain pathogenic diseases such as hepatitis C chronic inflammation is as much a factor in the disease manifestation as the virus itself (Nathan, 2002).',\n",
       " 'Also, incomplete or incorrect resolution of inflammation triggers fibrotic pathologies such as idiopathic pulmonary fibrosis (Wynn and Vannella, 2016).',\n",
       " 'Finally, chronic inflammation provides fuel for cellular neoplasms and is a major contributing factor in the formation of malignant tumours.',\n",
       " 'Indeed, immune cells are present in all tumours (Balkwill and Mantovani, 2001; Balkwill, Charles and Mantovani, 2005).   ',\n",
       " 'Cancer-related inflammation develops in multiple ways (Figure 0.1).',\n",
       " 'The so-called ‘extrinsic’ pathway describes instances, such as human papilloma virus (HPV) positive squamous cell carcinoma of the head and neck, where chronic inflammation downstream of innate recognition of an infectious agent creates an environment powered by cytokines and growth factors in which neoplastic clones can thrive (Coussens and Werb, 2002).',\n",
       " 'The ‘intrinsic’ pathway describes instances where oncogenic signalling stimulates production of soluble mediators that create a bespoke inflammatory tumour microenvironment (TME) (Mantovani et al., 2008).',\n",
       " 'A less understood pathway is therapy-induced',\n",
       " '17cancer-related inflammation which may depend on the therapeutic mechanism-of-action, the baseline tumour immune composition, as well as the tissue (Figure 0.1) (Greten and Grivennikov, 2019).    ',\n",
       " 'Cancer-related inflammation facilitates multiple cancer hallmarks such as angiogenesis, invasion, metastasis, survival and proliferation, yet equally both innate and adaptive inflammatory cells mount effective, tumour-specific immune responses (Greten and Grivennikov, 2019).',\n",
       " 'As such, cancer inflammation, albeit stemming from physiological processes that maintain homeostasis, nurtures or hinders the disease in distinct ways.',\n",
       " 'Major open questions for the cancer field are therefore how cancer-inhibitory inflammation goes awry, and the extent to which pro-tumourigenic inflammation contributes towards this (Shalapour and Karin, 2019; Hou, Karin and Sun, 2021).',\n",
       " 'Another open question relates to whether there are qualitatively different types of inflammation in human tumours or whether cancer inflammation is a consistent phenomenon that merely varies quantitatively.   ',\n",
       " 'Figure 0.1: Examples of pathways that account for chronic tumour inflammation.',\n",
       " 'The left panel shows how tissue injury associated with infectious agents for example leads to chronic inflammation which contributes to high levels of soluble factors that stimulate neoplasia and cancer-promoting inflammation.',\n",
       " 'The middle panel indicates how oncogenic mutations drive production of soluble mediators that sculpt a bespoke inflammatory TME.',\n",
       " 'The right panel shows how therapeutics induce cell death resulting in tumour control or tumour recurrence associated with unique inflammatory profiles.  ',\n",
       " 'Resolution of inflammation  A persistent dogma in the onco-immunology field is that cancers are wounds that fail to heal properly.',\n",
       " 'The wound healing response involves haemostasis, inflammation, angiogenesis and the generation of connective tissue which bears similarities to the formation of tumours (Dvorak, 1986).',\n",
       " 'This pattern of events is observed in cardiac tissue following myocardial infarction and in the skin after a laceration.',\n",
       " 'In tumours it was suggested that the resolution phase of the wound healing response  ',\n",
       " '18becomes perturbed through sustained inflammatory signalling, secretion of pro-angiogenic and invasive factors that promote extravasation (Balkwill and Mantovani, 2001).',\n",
       " 'Thus, rather than tumours creating a bespoke environment in which they can prosper, they instead rely on regular, predictable and physiological events that fuel progression (Dvorak, 1986).',\n",
       " 'This view now conflicts with many of the studies that link oncogenic signalling to the establishment of novel, reciprocal, inflammatory signalling circuits (Figure 0.2).',\n",
       " 'The most pre-eminent and prevailing view is now that for the majority of tumours, in which a chronic inflamed state does not precede tumour development, transformed cells hijack inflammatory signalling cascades to create tailored conditions in which they can prosper (Greten and Grivennikov, 2019).',\n",
       " 'This signalling establishes a forward feedback loop to recruit new haematopoietic cells, and induces local parenchymal cells to adopt heterogeneous states that are unique to the tumour site.',\n",
       " 'Transcriptional, proteomic and spatial heterogeneity is evident between healthy tissues, inflamed sites, primary tumours and metastatic locations (Greten and Grivennikov, 2019).',\n",
       " 'Through oncogene addiction and the accompanying metabolic re-wiring, cellular stress, cell death and secretion of myriad growth factors, cytokines and chemokines, malignant neoplasms create conditions more akin to an entirely new, inflamed tissue.',\n",
       " 'Some but not all bladder, liver, stomach, colon and pancreatic cancers emerge from states of chronic inflammation when the resolution phase of the inflammatory response goes awry.',\n",
       " 'Therefore, to understand how chronic inflammation fosters neoplasia, it is equally important to understand how  tissue homeostasis is restored in physiological settings (Serhan and Savill, 2005).',\n",
       " 'Eicosanoids are central mediators of this process.',\n",
       " 'Lipoxygenase (LOX) converts arachidonic acid into a series of lipoxins and leukotrienes whereas cyclooxygenase (COX) generates prostaglandins from the same precursor molecule.',\n",
       " 'These enzymes and eicosanoids are involved in a highly regulated, fine-tuned, and temporally sequenced set of interactions that orchestrate acute inflammation but also resolution (Levy et al., 2001).',\n",
       " 'Although cyclooxygenase-2 (COX-2) is the target of successful ‘anti-inflammatory’ drugs, prostaglandin-E2 (PGE2), that is downstream of COX-2, has a more nuanced role and these drugs are perhaps best considered as ‘immunomodulatory’.',\n",
       " 'In coordination with other eicosanoids PGE2 first aids trafficking of neutrophils to the source of the injury thereby promoting inflammation.',\n",
       " 'Later in the response, however, PGE2 facilitates a switch in neutrophil eicosanoid production favouring 15-lipoxygenase-2 products that are important in the resolution phase (Levy et al., 2001).',\n",
       " 'Therefore, the effect of PGE2 depends on the context, timing and phase of the inflammatory response (Levy et al., 2001; Nathan, 2002).',\n",
       " 'Indeed, this is certainly the case in cancer.',\n",
       " 'Dysregulation of these pathways ultimately facilitates tumour progression yet the cancer cell-intrinsic mechanisms that promote this dysregulation remain unclear, and the interaction of these lipid mediators with the diversity of intratumoural leukocyte and mesenchymal populations is incompletely understood (Wang and Dubois, 2010).',\n",
       " 'The increasing use of single cell RNA sequencing (scRNAseq) to measure expression of genes that encode for soluble proteins and their receptors has led to the development of new bioinformatics tools that infer ligand-receptor-cell state interaction networks (Browaeys, Saelens and Saeys, 2020).',\n",
       " 'These tools however overlook the activity of small lipid molecules such as PGE2 that have the potential to orchestrate various aspects of tumour biology and inflammation.',\n",
       " 'Thus, modern bioinformatic tools that facilitate discovery in heterocellular systems are not designed to leverage our knowledge of',\n",
       " '19biosynthetic pathways.',\n",
       " 'Altogether, there is a need to better understand how tumours hijack normal physiological processes to shape the inflammatory TME.',\n",
       " 'Our ability to successfully manipulate the inflammatory TME to deliver long-lasting anti-cancer immune responses has grown immeasurably since the advent of immune checkpoint blockade (ICB) drugs that block inhibitory receptors such as CTLA-4 and PD1 on activated T cells, or PD-L1 that ligates PD1 (Ribas and Wolchok, 2018).',\n",
       " 'Despite being transformative, these immunotherapies do not benefit all patients.',\n",
       " 'Understanding why is a key priority for the field (Hegde and Chen, 2020), and central to this is elucidating the complexity of cancer inflammatory phenotypes.',\n",
       " 'Before discussing ICB in depth, we will first describe the key features of cancer-related inflammation all of which potentially impact clinical ICB responses.  ',\n",
       " 'Cancer-related inflammation  Evidence that inflammation fosters cancer progression comes from a plethora of epidemiological studies.',\n",
       " 'Firstly, chronic inflammatory conditions predispose to several cancer types.',\n",
       " 'In the case of colorectal cancer for example, inflammatory bowel disease is a strong aetiological factor (Balkwill and Mantovani, 2001).',\n",
       " 'In addition, anti-inflammatory agents such as aspirin, a non-steroidal anti-inflammatory drug, and canakinumab, an IL-1β blocking antibody, were shown to reduce the risk of cancer development (Baron et al., 2003; Ridker et al., 2017).',\n",
       " 'Murine models of cancer have also elucidated several mechanisms that underpin tumour-promoting inflammation.',\n",
       " 'For instance, myeloid cell-specific deletion of IKKb which promotes signalling via NFκB transcription factors that switch on expression of inflammatory cytokines, led to a 50% reduction in tumour formation in a model of colitis-induced colon cancer (Greten et al., 2004).',\n",
       " 'Thus, tumour development was inhibited by disrupting the ability of myeloid cells to signal via NFκB. These findings show that immune cells are conscripted to provide additional fuel to the developing tumour.  ',\n",
       " 'As mentioned, tumour cells mould a bespoke inflammatory TME in part by pro-actively secreting cytokines and chemokines.  ',\n",
       " 'Many studies have uncovered mechanistic links between oncogenic signals and inflammatory gene expression by cancer cells (Figure 0.2).',\n",
       " 'These include MYC, TP53 and KRAS pathways amongst others (Spranger and Gajewski, 2018).',\n",
       " 'Despite these discoveries, doubt remains as to whether these pathways converge upon common inflammatory transcriptional programs or whether there are unique immunomodulatory characteristics of each.',\n",
       " 'Some evidence exists that different pathways collaborate to elevate inflammatory signalling, yet it is unclear how widespread this phenomenon is (Kortlever et al., 2017).',\n",
       " 'Furthermore, it is also poorly understood how genomic instability, aneuploidy and somatic copy number alterations (SCNAs) influence the ability of tumour cells to manipulate the inflammatory composition.',\n",
       " 'Notwithstanding the importance of uncovering the formational events of chronic inflamed malignancies, it is essential to further define what cancer-related inflammation actually is and the degree to which it is actively involved in the tumour development process.   ',\n",
       " '20 Figure 0.2: Examples of immunomodulatory oncogenic pathways.',\n",
       " 'Immunomodulatory oncogenic events include gain-of-function mutations in KRAS (Kumar et al., 2012; Liao et al., 2019; Li et al., 2020), LKB1 (Koyama et al., 2016), and the WNT pathway (Spranger, Bao and Gajewski, 2015; Avgustinova et al., 2016), loss-of-function mutations in PTEN (Peng et al., 2016; Xiao et al., 2018) and TP53 (Wellenstein et al., 2019; Blagih et al., 2020), FAK and MYC amplifications (Shchors et al., 2006; Jiang et al., 2016; Kortlever et al., 2017), loss of ATM (McGrail et al., 2018), and epigenetic dysregulation (Nishida et al., 2020; Gangoso et al., 2021).   ',\n",
       " 'Components of the inflammatory tumour microenvironment  Cancer inflammation consists not just of a cellular component but also an important extracellular component comprising tissue matrix proteins, enzymatic cascades, metabolites, lipids, cytokines, chemokines and antibodies (Figure 0.3).',\n",
       " 'A primary objective of this project was to define common features of the cancer inflammatory phenotype.',\n",
       " 'Although tissue resident immune subsets promote inflammation, and the local tissue architecture and metabolism can be important determinants of inflammatory composition, chronic states are largely dependent on de novo recruitment of immune cells (Medzhitov, 2008; Salmon et al., 2019).',\n",
       " 'Almost all of these tissue resident and newly recruited cell types have pleiotropic roles in cancer progression.   ',\n",
       " '21 Figure 0.3: Examples of the major components of the cancer inflammatory phenotype.',\n",
       " 'A figure illustrating examples of four key features of the inflammatory TME including clockwise from the top left the cell types that occupy the TME, extracellular cascades and aspects of the tissue architecture (ECM = extracellular matrix), features of the local TME impacting upon the systemic inflammatory response, and finally molecular features of the TME including small molecule and protein signalling factors.  ',\n",
       " 'Infiltrating T lymphocytes, for example, are a major cellular component of the tumour inflammatory composition thought to contribute towards endogenous and therapy-induced tumour immunity.',\n",
       " 'The vast majority of tumours have however evolved to co-exist with these cells.',\n",
       " 'Some tumours evolve ways of selectively recruiting regulatory T cells (Tregs) that perpetuate immune tolerance (Sakaguchi et al., 2008; Wang et al., 2019).',\n",
       " 'Other tumours render cytotoxic lymphocytes inert, via changes in metabolism, through upregulation of inhibitory checkpoint receptors, and the secretion of suppressive cytokines (Nakamura and Smyth, 2019).',\n",
       " 'These immune evasive features frequently alleviate the selective pressure to expel these cells from the tissue.',\n",
       " 'This subversion of infiltrating T cells is a characteristic of advanced cancers, yet cancer inflammation can be an early event.',\n",
       " 'Indeed, T cell activity early in tumour formation sculpts the formation of tumours with low immunogenicity through a Darwinian selection process known as immunoediting which results in extensive intratumoural heterogeneity (ITH) (Dunn, Old and Schreiber, 2004; McGranahan and Swanton, 2017; Dentro et al.,  ',\n",
       " '222021)',\n",
       " '.',\n",
       " 'For example, methylcholanthrene (MCA)-induced tumours in RAG1 or 2-deficient mice, that lack mature T cells and B cells, were highly immunogenic compared with tumours derived from wild-type (WT) mice (Shankaran et al., 2001).',\n",
       " 'Thus, T cells eliminate highly immunogenic clones and those clones with low immunogenicity and immune evasive features become dominant in the tumour and tolerate the presence of cytotoxic lymphocytes.',\n",
       " 'These cytotoxic lymphocytes then become established components of the inflammatory milieu and may even actively contribute towards tumour development and evolution.',\n",
       " 'In addition to T cells, neutrophil-deficient mice are also more susceptible to MCA-induced sarcomas (Ponzetta et al., 2019).',\n",
       " 'Therefore, phagocytes and innate immune pathways also contribute to the elimination phase of tumour immunoediting prior to their gradual skewing towards pro-tumoural states (Vesely et al., 2011).',\n",
       " 'Thus, inflammatory cells may have dual roles at different stages of the immune response to a developing tumour.',\n",
       " 'We will now discuss the types of non-tumour cell that occupy the TME and the ways in which these cells are believed to contribute towards both tumour progression and tumour immunity.  ',\n",
       " 'Lymphocytes Mature lymphocytes derive from common lymphoid bone marrow stem cell precursors.',\n",
       " 'T lymphocytes mature in the thymus and periphery once they leave the bone marrow.',\n",
       " 'Several major T cell subsets exist.',\n",
       " 'Conventional T cells have a T cell receptor (TCR) containing a and b chains, whereas gdT cells have productive g and d chains.',\n",
       " 'Unlike gdT cells, conventional T cells undergo stringent positive and negative selection in the thymus.',\n",
       " 'TCRs capable of binding to major histocompatibility (MHC) class I and II molecules with a particular range of affinities are selected for.',\n",
       " 'This process delineates CD4+ T cells, that bind MHC class II molecules, and CD8+ cells that bind MHC class I molecules (Germain, 2002).',\n",
       " 'Negative selection occurs in the thymus primarily by specialised antigen-presenting cells (APCs) that screen TCR: self-peptide-MHC interactions, and eliminate highly self-reactive clones.',\n",
       " 'Naïve and antigen-inexperienced T cells leave the thymus and circulate through lymphoid tissues (Germain, 2002).',\n",
       " 'In tissue-draining lymph nodes, productive T cell priming takes place by a contact dependent interaction with mature dendritic cells (DCs) and depends upon two main signals.',\n",
       " 'Firstly, the TCR must bind to peptide-bound MHC class II, and secondly, T cells must receive critical co-stimulatory signals (Chen and Flies, 2013).',\n",
       " 'A plethora of co-stimulatory and co-inhibitory receptors evolved in humans to fine-tune this process in order to mount proportionate immune responses depending on the challenge  (Chen and Flies, 2013).',\n",
       " 'Following antigen recognition and co-stimulation T cells become activated, proliferate, and differentiate to become effector cells.',\n",
       " 'Effector CD8+ T cells are cytotoxic and travel to sites of infection or inflammation to enact antigen-specific killing.',\n",
       " 'Activated helper CD4+ T cells likewise are recruited to sites of damage where their function depends on cell fate decisions taken earlier in development.',\n",
       " 'Both CD8+ and CD4+ T cells demonstrate broad and dynamic transcriptional states that differ between the blood, healthy and tumour tissue (Guo et al., 2018).     ',\n",
       " 'CD8+ T cells Cytotoxic CD8+ T cells are the effector cells of tumour immunity.',\n",
       " 'CD8+ T cell density, abundance and location in the tumour associates with prolonged patient survival (Galon et al., 2006;',\n",
       " '23Mahmoud et al., 2011; Fridman et al., 2012; Bindea et al., 2013).',\n",
       " 'Several CD8+ T cell transcriptional states have been identified in tumours that are not observed in the blood.',\n",
       " 'A common nomenclature for these CD8+ T cell states has been described (van der Leun, Thommen and Schumacher, 2020).',\n",
       " 'This characterisation places cells on a trajectory from naïve-like T cells through various and increasingly dysfunctional states, and divergently a population of cells with enhanced cytotoxicity.   ',\n",
       " 'In addition to their cytotoxic capacity, interferon-g (IFNg) is an important cytokine in the armoury of activated CD8+ T cells (Kaplan et al., 1998).',\n",
       " 'There are multiple mechanisms by which IFNg contributes towards host tumour-immunity including anti-proliferative, anti-angiogenic and pro-apoptotic activity (Ikeda, Old and Schreiber, 2002).',\n",
       " 'Critically, IFNg induces MHC class I expression by tumour cells that stimulates antigen presentation (Dighe et al., 1994).',\n",
       " 'Reflecting the importance of IFNg as a cytokine with anti-tumour activity, a six gene IFNg signalling signature predicted progression-free survival (PFS) benefit at baseline prior to treatment with immunotherapeutic antibodies (Ayers et al., 2017).',\n",
       " 'This suggests that pre-existing IFNg signalling is a common feature of the tumour inflammatory phenotype of immune checkpoint blockade (ICB)-sensitive patients.   ',\n",
       " 'Dysfunction, that is an inevitability for many infiltrating CD8+ T cells in advanced cancers, is a dynamic process, that in tumours, remains poorly understood.',\n",
       " 'The TME, cytokine signalling, lipid mediators, exposure to inhibitory checkpoint molecules, hypoxia and changes in tumour metabolism have all been reported to promote T cell dysfunction that correlates with increasing expression of receptors such as LAG3, ENTPD1, PDCD1 and the activity of transcription factors such as TOX (McLane, Abdel-Hakeem and Wherry, 2019; van der Leun, Thommen and Schumacher, 2020).',\n",
       " 'It remains unclear how both antigen-specific and bystander CD8+ T cells transition into, between and out of these dysfunctional states, and what the implications are for tumour immunity and cancer patient prognosis.  ',\n",
       " 'Mutated peptides are presented by MHC class I molecules on the surface of tumour cells.',\n",
       " 'CD8+ T cells with specificity to these neo-epitopes can deliver profound anti-tumour immune responses (DuPage et al., 2012).',\n",
       " 'Such antigen-specific CD8+ T cells are also important for ICB responses as well as endogenous immunity (Gubin et al., 2014).',\n",
       " 'T cells may also drive tumour immunity by targeting wild-type antigens (Lo et al., 2021).',\n",
       " 'As such, there are multiple ways by which CD8+ T lymphocytes are reported to elicit tumour immunity, but it remains unclear why patients with rich CD8+ T cell infiltrates can have dismal outcomes and whether their subversion is key to this.',\n",
       " 'CD8+ T cells make up only a proportion of the T lymphocytes that infiltrate tumours, with many tumours also infiltrated by CD4+ T cells.  ',\n",
       " 'CD4+ T cells CD4+ T helper cells have a TCR that binds peptide-MHC class II complexes on APCs such as DCs and macrophages.',\n",
       " 'CD4+ T cells are also capable of neo-epitope reactivity and implicated in neoantigen-specific ICB responses although they lack the cytotoxic capacity of CD8+ T cells (Linnemann et al., 2015; Alspach et al., 2019).',\n",
       " 'Nevertheless, CD4+ T cells are thought to augment the killing function of CD8+ T cells providing cytokine support via IL-2 and IFNg, as well as directly engaging activating receptors on the CD8+ T cell surface and dampening expression of inhibitory receptors',\n",
       " 'such',\n",
       " '24as PD1 (Ahrends et al., 2017).',\n",
       " 'Upon activation, these helper T cells express CD40L that ligates CD40 on the surface of APCs leading to upregulation of co-stimulatory receptors CD80 and CD86 on APCs, enhanced survival, and ultimately to more productive APC-T cell interactions as opposed to those that promote tolerance (Quezada et al., 2004).',\n",
       " 'CD40 agonists, mimicking the activity of helper T cells, are efficacious in pre-clinical cancer models and have entered clinical development  (Beatty et al., 2011; Hoves et al., 2018; Baumann et al., 2020; Vonderheide, 2020).',\n",
       " 'Ultimately, the importance of these cells in shaping the inflammatory TME and in therapy-induced tumour immunity remains ambiguous.  ',\n",
       " 'Regulatory T cells Tregs, as previously described, restrain immune responses and promote tolerance of self-antigens (Sakaguchi, 2005).',\n",
       " 'These CD25+FOXP3+CD4+ T cells have potent immunosuppressive functions characterised by IL-10, IL-35 and transforming growth factor β (TGFβ) production together with high CTLA-4 expression (Sakaguchi et al., 2008).',\n",
       " 'Indeed, the efficacy of CTLA-4 blocking antibodies at least partly depends on Treg depletion (Quezada et al., 2008; Wing et al., 2008; Peggs et al., 2009).  ',\n",
       " 'Targeting Tregs to raise tumour immunogenicity remains an attractive prospect.',\n",
       " 'For example, an anti-CD25 antibody, developed to selectively deplete Tregs whilst sustaining IL-2-STAT5 signalling in effector T cells, synergised with an anti-PD-L1 antibody and a cancer cell vaccine in syngeneic murine cancer models (Solomon et al., 2020).',\n",
       " 'Thus, novel ways of manipulating the T cell compartment either by selectively expanding anti-tumour effector T cells, or depleting Tregs, may improve pre-existing and ICB-induced anti-tumour immune responses (Sun et al., 2019).  ',\n",
       " 'γδT cells γδT cells are an unconventional T cell subset expressing a heterodimeric γδTCR, the antigen specificity of which is unknown.',\n",
       " 'These cells may promote metastasis through cytokine production and by inducing expansion of neutrophils (Coffelt et al., 2015).',\n",
       " 'In contrast, some γδT cells are believed to secrete IFNg and promote tumour immunity hence their potential pleiotropic role in cancer (Li et al., 2021).   ',\n",
       " 'Natural killer cells NK cells are innate lymphoid cells (ILCs) forming an essential component of the innate immune system.',\n",
       " 'These cells derive from a bone marrow precursor common to other ILC populations like ILC1, ILC2 and ILC3s (Chiossone et al., 2018).',\n",
       " 'ILC1-3 are less well studied in cancer biology than NK cells, although there are some reports that these cells may secrete cytokines such as IFNg, IL-22, IL-17 and IL-13 (Chiossone et al., 2018).  ',\n",
       " 'Unlike CD8+ T cells, NK cells do not have a TCR, cannot mount antigen-specific responses and are a rare intratumoural population.',\n",
       " 'Nonetheless, NK cells are broadly considered as purveyors of tumour immunity with potent cytotoxic activity, and the ability to secrete soluble factors such as TNFα, IFNg and XCL1 that have the capacity to shape the TME and even recruit conventional DCs (Barry et al., 2018; Böttcher et al., 2018; Chiossone et al., 2018; Bonavita et al.,',\n",
       " '252020).',\n",
       " 'As such, in addition to their direct killing capacity NK cells are capable of bridging innate and adaptive immunity.',\n",
       " 'In physiological conditions, NK cell activity is suppressed by the binding of MHC class I molecules to inhibitory receptors on the NK cell surface (KIR receptors)',\n",
       " '(Pegram et al., 2011; Chiossone et al., 2018).',\n",
       " 'NK cells dynamically express a plethora of activating and inhibitory receptors thereby enabling responses to be fine-tuned.',\n",
       " 'Tumours frequently downregulate MHC class I molecules to avoid antigen recognition by CD8+ T cells, and upregulate stress ligands such as ULBP1-3, MICA and MICB (Raulet and Guerra, 2009).',\n",
       " 'These events can activate NK cells, yet tumours have evolved several ways of specifically evading NK cell activity including the production of secreted proteins and lipids that inhibit NK cells and enzymes that cleave NK stimulatory receptors from the tumour cell surface (Ben-Shmuel, Biber and Barda-Saad, 2020; Sheffer et al., 2021).',\n",
       " 'NKG2D, encoded by the gene KLRK1, is an example of an activating receptor that establishes a threshold for activation early in development (Jelenčić et al., 2018).',\n",
       " 'KLRK1-deficient mice have impaired tumour surveillance in models of spontaneous prostate cancer (Guerra et al., 2008).',\n",
       " 'Paradoxically, in liver cancers driven by inflammation, NKG2D plays a pro-tumourigenic role facilitating hepatocyte proliferation, damage and an influx of memory CD8+ T cells (Sheppard et al., 2017).',\n",
       " 'Thus, NK cells and their surface receptors can have context-dependent and paradoxical roles in tumour progression.',\n",
       " 'Despite this, NK cells were enriched in peritumoural and intratumoural regions of melanomas that responded to ICB despite loss of MHC class I (Lee et al., 2018).',\n",
       " 'NK cells may therefore be implicated in ICB responses when tumours have evolved mechanisms to bypass antigen-specific T cell-mediated immunity.   ',\n",
       " 'Myeloid cells A considerable proportion of the immune cells in tumours are of myeloid lineage.',\n",
       " 'This includes the most abundant macrophages and monocytes, followed by the less prevalent granulocytes, DCs and mast cells.',\n",
       " 'Myeloid cells are highly plastic meaning exposure to different stimuli in the TME can profoundly alter their transcriptional state and they are able to transition between these states (Xue et al., 2014; Gubin et al., 2018).',\n",
       " 'In addition, myeloid cells are developmentally distinct with different subsets specialised for roles in peripheral tissues and in lymphoid organs (Guilliams et al., 2014; Veglia et al., 2021).',\n",
       " 'In recent years, considerable debate has focussed on the characterisation and nomenclature of tumour-infiltrating myeloid cell types confounded by widespread inconsistency in the use of functional and ontogeny-based definitions (Hegde, Leader and Merad, 2021).',\n",
       " 'Herein, we will focus on cell types with ontogeny-related definitions thereby excluding so-called myeloid-derived suppressor cells that are variably defined by surface phenotype and function ex vivo.  ',\n",
       " 'Macrophages The vast majority of tumour-associated macrophages (TAMs) derive from circulating monocytes that begin to differentiate into macrophages once they leave the circulation (Movahedi et al., 2010).',\n",
       " 'Therefore, monocytes with varying degrees of differentiation are detected in tumours (Chevrier et al., 2017).',\n",
       " 'Macrophages are phagocytes capable of clearing apoptotic or infected cells.',\n",
       " 'Although mature macrophages express high levels of MHC class II, and are capable of antigen presentation, they are not thought to contribute substantially towards tumour antigen-specific T cell',\n",
       " '26stimulation (Salmon et al., 2016).',\n",
       " 'Nevertheless, these phagocytes influence almost all aspects of tumour biology including proliferation, intravasation, angiogenesis and response to therapy (DeNardo and Ruffell, 2019).',\n",
       " 'The high plasticity of macrophages responding to diverse environmental cues means they are not a homogeneous population intratumourally, nor do they fall into binary subsets.',\n",
       " 'An historical and simplified view of macrophage biology categorises them into M1 and M2 subsets based on in vitro differentiation of monocytes in the presence of lipopolysaccharide or IFNg (M1) or IL-4 and IL-13 (M2).',\n",
       " 'However, this nomenclature overlooks the degree of phenotypic heterogeneity displayed by TAMs.',\n",
       " 'The macrophage transcriptional state is most dramatically affected by the fibrotic, metabolic and immune characteristics of the resident tissue, alongside its anatomy and pathology (DeNardo and Ruffell, 2019).',\n",
       " 'In parallel, macrophage dynamics are strongly influenced by diverse soluble factors and damage-associated molecular patterns (Xue et al., 2014; Gubin et al., 2018).',\n",
       " 'The full spectrum of macrophage phenotypes across different cancer types has been illuminated by scRNAseq, as well as proteomic approaches such as cytometry by time of flight (CYTOF) (Zilionis et al., 2019; Cheng et al., 2021).',\n",
       " 'Highlighting their complexity, macrophage transcriptional phenotypes can differentially correlate with tumour fate (Mantovani and Sica, 2010; DeNardo and Ruffell, 2019; Braun et al., 2021).',\n",
       " 'Through autocrine and paracrine signalling, the TME is capable of adapting to therapies that modulate monocyte-macrophage dynamics (Quail et al., 2016; Kumar et al., 2017).',\n",
       " 'This confirms the importance of macrophages towards tumour progression and evolution whilst exposing the difficulty in developing successful immunomodulatory agents (Mantovani et al., 2017).  ',\n",
       " 'A prevalent view in the field is that TAMs gradually acquire a more T cell-suppressive, pro-tumourigenic, and M2-like phenotype that progresses in line with tumour immunoediting (Mantovani and Sica, 2010; Dunn, Old and Schreiber, 2004).',\n",
       " 'At early stages of malignant transformation, M1-like TAMs contribute towards immunosurveillance through cytokine production and phagocytosis that is evaded by upregulation of do-not-eat-me signal CD47: a novel immunotherapy target (Jaiswal et al., 2009, 2010; Feng et al., 2019).',\n",
       " 'It remains to be seen whether CD47 blockade will however unleash the phagocytic capacity of TME-experienced TAMs.',\n",
       " 'By contrast, PI3Kγ was identified as a TAM intracellular molecular switch capable of downregulating genes implicated in innate immunity, antigen presentation and T cell activation whilst inducing expression of T cell suppressive factors such as TGFβ (Kaneda et al., 2016).',\n",
       " 'There is a strong rationale for identifying macrophage intracellular signalling nodes such as PI3Kγ that serve as points of signalling convergence with the power to skew differentiation in unfavourable directions.   ',\n",
       " 'Although blood-derived monocytes are a rich source of TAMs, tissue-resident macrophage populations  also influence the disease course (Davies et al., 2013).',\n",
       " 'These cells derive from progenitors that seed their respective tissues early in embryogenesis and gradually develop properties intrinsic to their residing tissue (Schulz et al., 2012).',\n",
       " 'Tissue-resident macrophages also self-renew locally throughout adulthood and are not supplemented by circulating monocytes (Hashimoto et al., 2013).  ',\n",
       " 'Kupffer cells in the liver, alveolar macrophages in the lung and FOLR2+ macrophages in the breast have been shown to regulate different aspects of tumour biology (Li et al., 2012; Gül et al., 2014; Casanova-Acebes et al., 2021; Ramos et al., 2021).',\n",
       " 'Whilst Kupffer cells were shown to restrain T cell activity via a galectin-9:TIM3 checkpoint interaction, FOLR2+ perivascular macrophages are thought to',\n",
       " '27promote CD8+ T cell-mediated anti-tumour immunity (Li et al., 2012; Ramos et al., 2021).',\n",
       " 'Therefore, highly specialised tissue-resident macrophages may equally support and restrain tumour immunity.  ',\n",
       " 'Neutrophils Unlike macrophages, neutrophils are short-lived phagocytes and the most abundant immune cell type in the blood.',\n",
       " 'Following infection or damage, neutrophils are rapidly recruited to the at-risk tissue following a chemokine gradient.',\n",
       " 'The release of CXC family (CXCL-1, -2, -5) chemokines as well as cytokines such as IL-1β, IL-6 and IL-8 by tissue-resident macrophages, mast cells, local parenchymal cells or indeed cancer cells, drives neutrophil recruitment to the tissue (Gregory and Houghton, 2011; Mantovani et al., 2011).',\n",
       " 'Like macrophages, neutrophils are critical cellular mediators of innate immunity, and possess phagocytic capacity that enables them to engulf pathogens.',\n",
       " 'Pre-formed granules, hence the term granulocytes, containing digestive enzymes such as elastase and myeloperoxidase are then safely deployed to fuse with the phagosome and kill the engulfed microorganism.',\n",
       " 'Interestingly, human, but not murine, neutrophils may retain their ability to selectively kill cancer cells (Cui et al., 2021).',\n",
       " 'There is however strong correlative evidence to suggest that neutrophils predominantly promote cancer progression in humans (Templeton et al., 2014; Gentles et al., 2015).',\n",
       " 'Thus, more often than not the anti-tumour properties of neutrophils are circumvented by tumours (Jaillon et al., 2020).',\n",
       " 'Indeed, the neutrophils most frequently detected in tumours display alternative features including PGE2 production, expression of nitric oxide synthase 2 (iNOS) and arginase 1 (Arg1), upregulation of PD-L1, TREM1, and production of cytokines such as IL-6 and IL-8 (Veglia et al., 2021).',\n",
       " 'These characteristics of intratumoural neutrophils contribute towards their potent suppressive effect towards tumour immunity, as well as their role in promoting metastasis (Coffelt et al., 2015; Szczerba et al., 2019; Wellenstein et al., 2019; Nishida et al., 2020; Teijeira et al., 2021).',\n",
       " 'It is therefore important to Identify the mechanisms by which neutrophils are subverted by cancer cells.',\n",
       " 'Notably, GM-CSF-mediated fatty acid transport protein 2 (FATP2) upregulation was shown to mediate this phenotypic switch upstream of PGE2 production by neutrophils (Veglia et al., 2019).',\n",
       " 'Targeting of nodes such as FATP2 may therefore relinquish the suppressive effect of neutrophils towards cytotoxic lymphocytes in the TME.  ',\n",
       " 'Dendritic cells DCs are developmentally heterogeneous.',\n",
       " 'Some DCs derive from monocytes, whereas conventional DCs (cDCs) derive from precursors upstream of monocytes during haematopoiesis (Guilliams et al., 2014).',\n",
       " 'Characterised by a spindle shape, DCs are morphologically distinct from macrophages and have functional differences that are particularly evident in cancer.  ',\n",
       " 'A number of DC subsets are observed in vivo, in the circulation, lymph nodes and the local TME (Villani et al., 2017; Zhang et al., 2020).',\n",
       " 'Broadly, cDCs can be separated into three main classes.',\n",
       " 'Plasmacytoid DCs (pDCs) are rarely found to infiltrate tumours.',\n",
       " 'cDC1s efficiently process tumour antigens released by dying cancer cells, migrate to secondary lymphoid organs and present tumour antigens to both CD4+ and CD8+ T cells thereby potently inducing T cell priming and activation (Broz et al., 2014; Roberts et al., 2016; Ferris et al., 2020).  ',\n",
       " 'These cells may also promote T cell infiltration by secreting chemoattractants',\n",
       " '28CXCL9 and CXCL10 (Spranger et al., 2017).',\n",
       " 'As such, cDC1s are considered as essential components of a cancer-inhibitory inflammatory phenotype, such that tumours have evolved several mechanisms to subvert their activity (Spranger, Bao and Gajewski, 2015; Giampazolias et al., 2021; Haas et al., 2021).',\n",
       " 'Therapies that stimulate DC maturation and expansion may overcome these barriers (Salmon et al., 2016; Haas et al., 2021).  ',\n",
       " 'A second subset of cDCs called cDC2s expresses CD11b, and has distinct antigen processing behaviour and cytokine expression than cDC1s (Merad et al., 2013).',\n",
       " 'cDC2s are less efficient at cross-priming T cells than cDC1s, are thought to be specialised for CD4+ T cell priming in lymph nodes, and promoting T follicular helper cell (TFH) differentiation (Gutiérrez-Martínez et al., 2015;',\n",
       " 'Krishnaswamy et al., 2017).',\n",
       " 'Finally, a fourth population of cDCs that infiltrates tumours (referred to herein as cDC3) has more recently been described following extensive single cell analysis of human and mouse tumours.',\n",
       " 'These cells express high levels of CCR7 and in humans LAMP3 (Cheng et al., 2021).',\n",
       " 'It is more likely that cDC3s represent a cDC activation state rather than a separate DC lineage (Gerhard et al., 2020).',\n",
       " 'Interestingly, recent reports suggest that cDC3s express high levels of PD-L1, endowing them with the potential to regulate effector PD1+ T cells (Maier et al., 2020).',\n",
       " 'These findings suggest that DCs can have a dual role in anti-tumour immune responses.   ',\n",
       " 'Mesenchymal cells  ',\n",
       " 'In addition to leukocytes and cancer cells themselves, mesenchymal cells are important components of the tumour milieu capable of secreting soluble inflammatory mediators.',\n",
       " 'Several mesenchymal or stromal cell subsets exist in tumours including pericytes, blood and lymphatic endothelial cells, and cancer-associated fibroblasts (CAFs) (Hutton et al., 2021).',\n",
       " 'CAFs are a major, abundant subset of tumour stromal cells that contribute towards diverse cancer hallmarks (Sahai et al., 2020).',\n",
       " 'The current belief is that CAFs can be sub-divided into two main groups; myofibroblasts and inflammatory CAFs (iCAF) (Öhlund et al., 2017).',\n",
       " 'Similar to the role of activated fibroblasts during the wound healing response, myofibroblasts are contractile, secrete ECM proteins and enzymes such as lysyl oxidase, as well as pro-fibrotic signalling molecules such as TGFβ.',\n",
       " 'These actions collectively drive fibrosis as well as epithelial-to-mesenchymal transition (EMT) (Sahai et al., 2020).',\n",
       " 'TGFβ itself is considered as a master regulator of fibrosis but also an immunosuppressive cytokine that has been shown to engage in inflammatory, autocrine feedback loops for example with IL-6 (Aoki et al., 2006; Mariathasan et al., 2018; Buechler, Fu and Turley, 2021).',\n",
       " 'By contrast, iCAFs are speculated to be a major source of several inflammatory mediators such as IL-1a, CSF1, CCL2 and IL-6 in human tumours (Buechler, Fu and Turley, 2021).',\n",
       " 'How to define, and distinguish CAF subsets is challenging.',\n",
       " 'CD105, for example, distinguishes two distinct CAF lineages that both express iCAF and myofibroblast genes in vivo (Hutton et al., 2021).',\n",
       " 'These subsets can be manipulated towards either state with recombinant proteins in vitro (Hutton et al., 2021).',\n",
       " 'Thus, CAFs exhibit plasticity and heterogeneity that is a product of their development, and of exposure to local environmental cues.',\n",
       " 'There is increasing evidence that CAFs interact closely with immune cells to shape the inflammatory composition of tumours in space and time (Sahai et al., 2020; Buechler, Fu and Turley, 2021).',\n",
       " 'The contribution of these cells towards pro-tumourigenic inflammation is still however up for debate.',\n",
       " '29Other cell types  B cells, mast cells, endothelial cells and tissue-specific cell populations also reside in the TME.',\n",
       " 'Increasing evidence links intratumoural B cell abundance, and tertiary lymphoid structures (TLSs) to anti-tumour immune responses (Cabrita et al., 2020; Helmink et al., 2020), although the mechanistic basis for this correlation is so far unclear.',\n",
       " 'Tumour endothelial cells are believed to promote the formation of TLSs and are capable of shaping various other features of the immune TME (Nagl et al., 2020).',\n",
       " 'The prognostic value of individual cell types varies greatly across different cancers as meta-analyses have shown (Bindea et al., 2013; Gentles et al., 2015; Charoentong et al., 2017).',\n",
       " 'This partly reflects the phenotypic plasticity of immune cells within the spatiotemporal dynamics of the tumour.',\n",
       " 'Monocytes, for example, are not universally associated with poor prognosis and disease progression (Charoentong et al., 2017).',\n",
       " 'Although these cells differentiate into tumour-promoting macrophages, patrolling monocytes may also control metastasis and promote NK cell activity (Hanna et al., 2016).',\n",
       " 'Inflammatory cell types can therefore have complex relationships with disease progression and prognosis.',\n",
       " 'A central question of this PhD thesis is whether or not there are common phenotypic states and molecular patterns that underpin the association of different leukocytes with patient survival and response to treatment rather than presence or absence of such cells.  ',\n",
       " 'Non-cellular components of the inflammatory TME  Inflammation provokes widespread changes to the cancerous tissue itself.',\n",
       " 'Aside from the cellular component of the inflammatory TME, the extracellular space comprises a complex network of matrix proteins and enzymes that are dysregulated in cancer.',\n",
       " 'Cancer cells, stromal cells, and immune cells produce laminins, collagens, cathepsins, matrix metalloproteinases (MMPs), protease inhibitors and other such ECM proteins.',\n",
       " 'Rather than a complete degradation of the ECM, inflammatory cells and cancer cells alike restructure, remodel and reorganise the ECM to favour invasion and intravasation (Greten and Grivennikov, 2019).',\n",
       " 'For example, in a model of skin carcinogenesis, host inflammatory cells were recruited to the TME and  conscripted to secrete MMP9 in aid of tumourigenesis (Coussens et al., 2000).',\n",
       " 'Therefore, immune cell interactions with the ECM may be critical in the construction of a pro-tumourigenic inflammatory TME.  ',\n",
       " 'Complement proteins are also present in tumours.',\n",
       " 'The complement system is involved in activation of innate immunity through humoral, rather than cellular mechanisms, and is a key component of the host defence against invading pathogens, that is also exploited by progressive tumours (Medler et al., 2018).',\n",
       " 'The complement cascade consists of a series of proteins and glycoproteins secreted by liver hepatocytes that can be activated in an antibody-dependent or independent manner to trigger recruitment of inflammatory cells, phagocytosis and direct attack of bacterial cell walls (Medzhitov, 2008).',\n",
       " 'The incidence of carcinogen-induced sarcomas was shown to be lower in mice lacking the complement protein C3 (Bonavita et al., 2015).',\n",
       " 'C3-deficiency was associated with lower levels of macrophage infiltration, suggesting an essential role for the complement system in mediating pro-tumourigenic inflammation (Bonavita et al., 2015).  ',\n",
       " '30Platelets are another component of the cancer inflammatory milieu.',\n",
       " 'Disruption of the vascular endothelium during tissue injury or carcinogenesis can lead to platelet infiltration of extracellular spaces which triggers the complement system and immune infiltration.',\n",
       " 'Platelets may even secrete modulatory cytokines such as TGFβ (Haemmerle et al., 2018) and mediate intravascular cancer cell protection, endothelial cell activation, seeding, and recruitment of myeloid cells to circulating tumour cell clusters, as well as metastatic niches (Lucotti et al., 2019).',\n",
       " 'Aspirin, a non-selective COX inhibitor, reversed these pro-metastatic effects by targeting the cyclooxygenase-1 (COX-1)/thromboxane-A2 (TXA2) axis in platelets (Lucotti et al., 2019).',\n",
       " 'These non-cellular components of inflammation confirm how diverse physiological inflammatory processes are re-purposed by developing tumours to aid progression and metastasis.',\n",
       " 'Cancer inflammation is therefore an exceedingly complex phenomenon characterised by extracellular processes, as well as cellular diversity.',\n",
       " 'Amidst this diversity a central aim of this project is to elucidate consistent cancer progressive and restrictive molecular traits and interactions.  ',\n",
       " 'Master regulators of cancer inflammation Identifying and understanding the master regulators of the tumour inflammatory milieu is important for the discovery of improved immunotherapeutics.',\n",
       " 'A master regulator of cancer inflammation can be defined as a cancer cell-intrinsic pathway or process that profoundly alters or shapes the aforementioned inflammatory properties of a tumour.',\n",
       " 'These encompass families of closely related proteins such as CXC family members and WNT ligands, as well as intracellular enzymes like COX-2 and processes such as apoptotic cell death (Table 0.1).',\n",
       " 'Such master regulators may in some cases overlap and interplay to sculpt the tumour immune composition.',\n",
       " 'For example, TGFβ itself is pro-fibrotic, and is believed to promote physical exclusion of T cells from the TME by modulating CAF dynamics, yet was also shown to induce cancer and stromal cell production of LIF that inhibits macrophage CXCL9 expression to prevent T cell infiltration, thus providing a double hit to ensure T cell expulsion (Albrengues et al., 2014; Mariathasan et al., 2018; Pascual-García et al., 2019).',\n",
       " 'Considering this level of interplay between potent regulators of the tumour inflammatory phenotype it is paramount to identify commonalities amongst them and to determine if specific pathways govern patient responses to different treatment regimens.',\n",
       " 'Master regulator Upstream Downstream effector Downstream effect Downstream reference CXC family chemokines BET bromodomain containing protein CXCL-1/IL-8',\n",
       " 'Neutrophil-dependent metastasis (Nishida et al., 2020)',\n",
       " 'NA IL-8',\n",
       " 'Circulating and tumour-infiltrating monocytes and neutrophils (Schalper et al., 2020) KRAS CXCR2 ligands Secretory cancer-associated fibroblasts with NFκB activation (Awaji et al., 2020) HRAS IL-8 Angiogenesis (Sparmann and Bar-Sagi, 2004)',\n",
       " 'Clonal effects CXCL1 Lack of T cell inflammation (Li et al., 2018) TGFb NA TGFβ',\n",
       " 'Fibroblast modulation, collagen deposition and T cell exclusion (Mariathasan et al., 2018)',\n",
       " 'NA TGFβ NK cell suppression via NKG2D downregulation (Lee et al., 2004)',\n",
       " '31BRAF, SMAD4, TP53 mutations, MYC amp TGFβ ECM remodelling (Chakravarthy et al., 2018) WNT ligands/ bcatenin APC, CTNNB1, APC2, AXIN1 βcatenin Lack of T cell inflammation (Gajewski et al., 2019)',\n",
       " 'BRAF mutant, PTEN loss βcatenin Lack of cDC1',\n",
       " 'and T cell infiltration (Spranger, Bao and Gajewski, 2015)',\n",
       " 'TP53 loss Several WNT ligands Macrophage IL-1b, systemic inflammation, neutrophil expansion and metastasis (Wellenstein and de Visser, 2018)',\n",
       " 'TLR4 overexpression βcatenin Neoplasia (Santaolalla et al., 2013) COX EPHA2-TGFβ COX-2 T cell exclusion (Markosyan et al., 2019)',\n",
       " 'NA COX-2/PGE2 EP2/4-mediated NK cell suppression (Bonavita et al., 2020)',\n",
       " 'NA COX-2/PGE2 cDC1 depletion (Böttcher et al., 2018)',\n",
       " 'NA COX-2/PGE2 Angiogenesis and metastasis (Xu et al., 2014)',\n",
       " 'NA COX-1/TXA2 Platelet-aided metastasis (Lucotti et al., 2019)',\n",
       " 'Vascular endothelium growth factor (VEGF)',\n",
       " 'NA VEGF Tumour diversity, reprogramming of the TME and T cell dysfunction (Ma et al., 2019)',\n",
       " 'NA VEGF Stifles DC maturation (Gabrilovich et al., 1996)',\n",
       " 'NA VEGF Macrophage polarisation (Huang et al., 2012)',\n",
       " 'LIF STAT3 LIF Suppression of CXCL9 production by macrophages and T cell exclusion (Pascual-García et al., 2019) TGFβ LIF Promotes pro-invasive fibroblasts (Albrengues et al., 2014) Interferons CDK4/6 inhibition Type III IFN Tumour antigen presentation (Goel et al., 2017) STAT1 Type I/II IFN Expression of T cell inhibitory receptors and severe T cell exhaustion due to chronic IFN signalling (Benci et al., 2019)',\n",
       " 'Apoptotic cell death Doxorubicin-treatment Caspase Phagocytosis (Casares et al., 2005)',\n",
       " 'RIPK1/ NFkB DAMPs DC maturation, cross priming and anti-tumour immunity (Nader et al., 2015)',\n",
       " 'Mitochondrial membrane permeabilisation/ NFkB TNF⍺ Macrophage activation with CXCL9/10 expression (Giampazolias et al., 2021)',\n",
       " 'ER stress High fat diet TNF⍺ Accrual of inflammatory macrophages (Nakagawa et al., 2014)',\n",
       " 'Unfolded protein response Core Cancer Secretome ECM modulation (Robinson et al., 2019)',\n",
       " 'Table 0.1:',\n",
       " 'Master regulators of the tumour inflammatory composition.  ',\n",
       " 'Immune checkpoint blockade Previous immunotherapies like high doses of IL-2 cytokine were marred by infrequent responses, inefficient immune activation and toxicities.',\n",
       " 'Immune checkpoint inhibitors have, however, galvanised the field of immunotherapy with market approval of CTLA-4, PD1 and PD-L1 targeted antibody drugs across several tumour types (Ribas and Wolchok, 2018).',\n",
       " 'Blocking additional checkpoints including LAG3 and TIGIT may also provide clinical benefit to patients, although this data is less mature (Rodriguez-Abreu et al., 2020; Lipson et al., 2021).',\n",
       " 'CTLA-4 was the first immune checkpoint receptor to be identified.',\n",
       " 'Its negative regulatory role in T cell activation was discovered when CTLA4-deficient mice developed lethal lymphoproliferative',\n",
       " '32disorders (Tivol et al., 1995).',\n",
       " 'Initially, CTLA-4 expressed by activated T cells was believed to suppress CD28-mediated co-stimulation of the TCR pathway by competing for binding to CD80 and CD86, thereby facilitating suppression of T cell activation and expansion (Linsley et al., 1991; Thompson and Allison, 1997; Pentcheva-Hoang et al., 2004).',\n",
       " 'Later, experiments revealed that the lymphoproliferative disorder that afflicts CTLA4-deficient mice, and the efficacy of CTLA-4 blockade in mouse cancer models, was dependent on the depletion of Tregs that constitutively express the receptor (Read, Malmström and Powrie, 2000; Takahashi et al., 2000; Quezada et al., 2008; Wing et al., 2008; Peggs et al., 2009; Simpson et al., 2013).',\n",
       " 'As such, there has been an evolution in the understanding of how anti-CTLA-4 therapy elicits clinical responses (Sharma et al., 2018).',\n",
       " 'Nevertheless, ipilimumab was the first anti-CTLA-4 drug approved for clinical use demonstrating for the first time that patient benefit could be derived through the use of monoclonal antibodies against checkpoint receptors (Hodi et al., 2010).',\n",
       " 'The next wave of clinical successes came with the approval of PD1 and PD-L1 targeting antibodies (PD(L)1).',\n",
       " 'Mice genetically ablated of PD-L1 survive, but suffer from increased susceptibility to experimental autoimmune encephalomyelitis (Latchman et al., 2004).',\n",
       " 'This reflects the negative regulatory role of PD-L1 towards PD1+ T cells.',\n",
       " 'In the context of cancer, PD-L1 is expressed both by tumour cells and myeloid cells (Taube et al., 2012).',\n",
       " 'Ligation of PD1 on activated T cells initiates signalling that suppresses proliferation, cytokine secretion and effector function (Freeman et al., 2000).',\n",
       " 'Tumour-specific, activated T cell clones are frequently identified within the TME yet upregulation of PD-L1, downstream of IFNg, avoids immune pathology (Spranger et al., 2013).',\n",
       " 'This mechanism of adaptive immune resistance has been exploited clinically (Taube et al., 2012; Tumeh et al., 2014).',\n",
       " 'Numerous clinical trials have demonstrated long-lasting patient benefits following inhibition of the PD1:PD-L1 axis for example in melanoma and bladder cancer (Hamid et al., 2013; Powles et al., 2014).',\n",
       " 'Since the approval of antibodies against PD(L)1, several studies have advanced our understanding of the mechanism by which these drugs work.',\n",
       " 'For example, macrophages that express PD1, and DCs expressing PD-L1, may have a role (Gordon et al., 2017; Dammeijer et al., 2020;',\n",
       " 'Oh et al., 2020; Strauss et al., 2020).',\n",
       " 'Paired scRNAseq and TCR sequencing (TCRseq) also showed that PD1 blockade promotes expansion of novel clonotypes in the TME, as opposed to reinvigorating pre-existing, tumour-infiltrating clones (Yost et al., 2019).',\n",
       " 'As such, the activity of checkpoint inhibitors in the periphery may be as important as inside the tumour.',\n",
       " 'Interactions between cell types in the TME, such as NK cells and cDC1s, may also be important factors in ICB responses highlighting the importance of the intercellular signalling network (Barry et al., 2018; Garris et al., 2018).',\n",
       " 'In summary, a new age of cancer immunotherapy has emerged that originates from the efforts of several research teams to tackle basic research questions about how the immune system is regulated.',\n",
       " 'The importance of the above-described seminal discoveries was highlighted by the award of a Nobel Prize to James Allison and Tasuku Honjo both of whom pioneered the early discovery checkpoint receptors.',\n",
       " 'Despite the overwhelming success of ICB several important questions remain (Hegde and Chen, 2020).',\n",
       " 'One fundamental unanswered question is how to predict which patients will benefit from treatment.',\n",
       " 'Although, tumour responses have been observed across several cancer types, many patients do not derive benefit (Topalian et al., 2012).',\n",
       " 'Only rare patients have the required tumour immunobiology and clinical characteristics to achieve clinical benefit, but there remains a lack of',\n",
       " '33agreement on the exact nature of what those characteristics are (Blank et al., 2016).',\n",
       " 'There is therefore an urgent requirement to clarify common mechanisms of sensitivity and resistance across tumour types, as well as to develop predictive biomarkers.',\n",
       " 'By identifying mechanisms of resistance to ICB, and thereby potential biomarkers, it is believed that new therapeutic targets will be discovered.',\n",
       " 'Indeed, the advent of ICB has led to an explosion in the number of clinical trials combining targeted therapies, radiotherapy, chemotherapies and other therapeutic modalities with ICB (Patel and Minn, 2018).',\n",
       " 'Defining combination strategies with the best chance of success, and coupling of these treatments to the appropriate patient population, are additional major challenges for the field that rely on the discovery of better predictive biomarkers (Corrales et al., 2016).',\n",
       " 'Biomarkers for immune checkpoint blockade  Identification of novel predictive biomarkers of response to ICB will ultimately serve to reduce side effects, healthcare costs and maximise benefit for patients (Havel, 2019).',\n",
       " 'Accurately estimating the likelihood of a patient responding to ICB is, however, a challenging prospect.',\n",
       " 'Thus far, measurement of PD-L1 expression by IHC, tumour mutational burden (TMB) and microsatellite instability (MSI) or deficient mismatch repair (dMMR) status have been approved as biomarkers for ICB (Topalian et al., 2016; Le et al., 2019; Marabelle et al., 2020).',\n",
       " 'Progress has therefore been made in delivering single modality biomarkers directly measuring the molecular target axis of ICB (PD-L1), and measuring surrogates of antigenicity that imply tumours with a greater propensity to elicit antigen-specific immune responses (TMB, MSI/dMMR).',\n",
       " 'There are however numerous other reported biomarkers that can be categorised into 4 main groups (Litchfield et al., 2021): (1) markers of T cell inflammation such as gene signatures of IFNg signalling, (2) sources of antigenicity such as TMB, (3) host factors, and (4) genomic drivers of immune escape like SCNA burden (Table 0.2).',\n",
       " 'Biomarker class Biomarker Measure Cancer type Association Reference T cell inflammation-related T cell-inflamed GEP (TIS) Nanostring Melanoma, HNSC, several other Sensitivity (Ayers et al., 2017) IFN𝛾 signalling Sensitivity CD8+ T cells IHC Melanoma Sensitivity (Tumeh et al., 2014) Cytolytic (CYT)',\n",
       " 'RNAseq Melanoma Sensitivity (Rooney et al., 2015) PD-L1 IHC Melanoma Sensitivity (Tumeh et al., 2014) IMPRES RNAseq Melanoma Sensitivity (Auslander et al., 2018)',\n",
       " 'TLS IHC Melanoma Sensitivity (Helmink et al., 2020)',\n",
       " 'RNAseq Melanoma Sensitivity (Cabrita et al., 2020) TIL proportion IHC Bladder Sensitivity (Snyder et al., 2017)',\n",
       " 'TCR repertoire TCRseq Melanoma Sensitivity (Tumeh et al., 2014)',\n",
       " 'ImmunoPhenoScore RNAseq Melanoma Sensitivity (Charoentong et al., 2017)',\n",
       " 'Antigenicity TMB WES Melanoma Sensitivity (Snyder et al., 2014) WES Bladder Sensitivity (Snyder et al., 2017)',\n",
       " 'WES NSCLC Sensitivity (Rizvi et al., 2015)',\n",
       " '34WES Melanoma Sensitivity (Van Allen et al., 2015) Clonal TMB WGS NSCLC, Melanoma Sensitivity (McGranahan et al., 2016)',\n",
       " 'Neoantigen burden WES Melanoma Sensitivity (Van Allen et al., 2015)',\n",
       " 'Neoantigen fitness WES Melanoma, NSCLC Sensitivity (Łuksza et al., 2017) UV induced mutations WES Merkel cell carcinoma Sensitivity (Knepper et al., 2019)',\n",
       " 'Tobacco induced mutations WES NSCLC Sensitivity (Anagnostou et al., 2020)',\n",
       " 'Clinical and germline factors Sex: Male Clinical Melanoma, NSCLC, HNSC, ccRCC, gastric, bladder Sensitivity (Conforti et al., 2018)',\n",
       " 'Tumour size Radiology Melanoma Resistance (Joseph et al., 2018)',\n",
       " 'HLA supertype B44 WES Melanoma Sensitivity (Chowell et al., 2018)',\n",
       " 'HLA supertype B62 WES Melanoma Resistance Genomic events RTK WES NSCLC Resistance (Anagnostou et al., 2020) SCNAs WES Melanoma Resistance (Davoli et al., 2017)',\n",
       " 'BRCA2 WES Melanoma Sensitivity (Hugo et al., 2016) JAK1/2 WES Melanoma Resistance (Zaretsky et al., 2016)',\n",
       " 'WES Melanoma, MMR-deficient colon Resistance (Shin et al., 2017)  ',\n",
       " 'Table 0.2: Examples of putative biomarkers of response to ICB.    ',\n",
       " 'Many of these putative biomarkers (Table 0.2) lack independent validation and benchmarking.',\n",
       " 'Moreover, the majority are directionally associated with ICB sensitivity, rather than resistance.',\n",
       " 'TMB, clonal TMB, neoantigen burden (NAB), neoantigen fitness, DNA damage response (DDR) pathway mutations, UV and tobacco signatures, clonal TCR repertoire, CD8+ T cells and T cell inflammation signatures all correlate with an increased chance of responding to ICB (Tumeh et al., 2014; McGranahan et al., 2016; Ayers et al., 2017; Keenan, Burke and Van Allen, 2019).',\n",
       " 'Biomarker discovery has been conceptually focused on the cancer immunity cycle and detecting elements of this linear path that contribute towards greater tumour immunogenicity (Chen and Mellman, 2013).',\n",
       " 'Even efforts to take a more unbiased view of the TME fail to give considerable weight to resistance-associated inflammatory characteristics (Charoentong et al., 2017).',\n",
       " 'The ImmunoPhenoScore, for example, is a multi-faceted biomarker comprising >75% variables related to antigen presentation and the T cell infiltrate (Charoentong et al., 2017).',\n",
       " 'It was reported to predict ICB response (Charoentong et al., 2017) but later demonstrated poor generalisability (Jiang et al., 2018).',\n",
       " 'As previously described in detail, the TME consists of a plethora of cell types, not just T cells, that can interact with one another, and with transformed cells, to influence tumour biology (Greten and Grivennikov, 2019).',\n",
       " 'Identifying common molecular entities that differentiate TMEs hostile to ICB might therefore be a fruitful approach in discovering more robust, multi-parametric biomarkers that incorporate hallmarks of resistance.',\n",
       " 'Several gene signatures of TME processes, signalling pathways and cell types that are thought to promote resistance to ICB have been reported (Figure 0.4).',\n",
       " 'In addition, whole transcriptome-based algorithms have demonstrated potential as predictive biomarkers compared to other more simplistic models, however difficulties surrounding the translation of RNAseq as a method into clinical practice may',\n",
       " '35ultimately limit the applicability of such methods (Jiang et al., 2018).   ',\n",
       " 'It is anticipated that improved accuracy in ICB response predictions will be achieved through combining biomarkers of different modalities.',\n",
       " 'The favoured of these approaches so far involves co-measurement of TMB and either T cell inflammation using the 18-gene Tumour Inflammation Signature (TIS), or CXCL9 expression alone (Cristescu et al., 2018; Litchfield et al., 2021).',\n",
       " 'It is unclear whether further improvements will come from measuring additional tumour genomic and transcriptomic features, host genetic factors or clinical and histopathological features of the disease.    ',\n",
       " 'Figure 0.4: Examples of TME characteristics associated with resistance to ICB or with poor prognosis.',\n",
       " 'These include, reading across from the top row, tumour hypoxia (Buffa et al., 2010), cancer cell proliferation (Pabla et al., 2019), EMT associated with the abundance of non-haematopoietic stromal cells (Wang et al., 2019), TGFb signalling on fibroblasts (Mariathasan et al., 2018) or associated with the ECM (Chakravarthy et al., 2018), angiogenesis and neo-vascularisation (Mcdermott et al., 2018), abundance of different immune and non-immune cell types (Danaher et al., 2017; Jiang et al., 2018), WNT and bcatenin signalling (Gajewski et al., 2019), a cancer cell-intrinsic immunotherapy resistance transcriptional program (Jerby-Arnon et al., 2018) and the transition from chronic inflammation to cancer associated with angiogenesis and ECM related factors (Chen et al., 2019).   ',\n",
       " 'Computational methods of interrogating the tumour inflammatory composition and discovering novel biomarkers  To study the role of tumour-infiltrating leukocytes in human cancer, historical studies harnessed pathology techniques.',\n",
       " 'Formalin-fixed paraffin embedded (FFPE) tumour tissue stained with haematoxylin and eosin (H&E) reveals cell nuclei (dark) and cytoplasm (pink), under the microscope.',\n",
       " 'By this approach immune cells with distinct morphology and nuclei density can be visualised and  ',\n",
       " '36quantified in tumours using rudimentary techniques (counting) or using modern day digital pathology methods that harness deep learning (Saltz et al., 2018).',\n",
       " 'Tumour-infiltrating lymphocytes (TILs), discernible by H&E, were shown to correlate with lymphocytes measured in tumour cell suspensions confirming the accuracy of pathological detection (Svennevig and Svaar, 1979).',\n",
       " 'Measurement of TILs by H&E was subsequently shown to be prognostic in some cancers, which stimulated efforts to standardise their quantification and characterisation resulting in the modern pathology guidelines for TIL estimation that exist today as a useful tool in the armoury of researchers studying cancer-related inflammation (Salgado et al., 2015; Hendry et al., 2017).  ',\n",
       " 'To characterise the immune composition of tumours at a deeper level IHC assays were developed using antibodies against leukocyte cell surface markers.',\n",
       " 'For example, CD20, CD3, CD4, CD8, CD45RO, CD68 and CD57-targeted antibodies enabled visualisation of the location and abundance of B cells, T helper cells, cytotoxic CD8+ T cells, memory T cells, macrophages and NK cells thereby increasing the granularity with which researchers could study immune populations in the TME (Negus et al., 1997).',\n",
       " 'By integrating gene expression profiles with IHC data for CD3, CD8, GZMB and CD45RO it was shown that quantitative and qualitative assessment of infiltrating T cells in the tumour core, and at the invasive margin, predicts disease-free survival (DFS) in colorectal cancer patients with greater power than disease staging criteria (Galon et al., 2006).',\n",
       " 'This seminal work set the stage for expanded IHC studies in which developmentally distinct immune cell types correlated in opposing directions with DFS in multiple cancer types (Bindea et al., 2013; Bruni, Angell and Galon, 2020).',\n",
       " 'Ultimately, these IHC studies helped shape the modern day understanding of the pleiotropic nature of inflammatory cells in tumours.  ',\n",
       " 'Next generation sequencing   Rather than measuring a limited, but spatially resolved, set of parameters, it is possible to generate measurements on a genome and transcriptome-wide scale.',\n",
       " 'The human genome project, painstakingly sequenced the entire human genome over a 15 year period using predominantly hierarchical shotgun sequencing (Lander et al., 2001).',\n",
       " 'This led to rapid advances in sequencing technology culminating in what is now referred to as next generation sequencing (NGS).',\n",
       " 'In parallel, the development of computational tools for genome alignment, de novo assembly, mutation calling, detection of copy number aberrations, aneuploidy and viral insertions among others has transformed efforts to understand cancer biology (Li and Durbin, 2009; Dobin et al., 2013; Goodwin, McPherson and McCombie, 2016; Cortés-Ciriano et al., 2021).',\n",
       " 'The International Cancer Genome (ICGC) and The Cancer Genome Atlas (TCGA) consortia harnessed the power of NGS to create extensive and unparalleled resources for studying more than 30 tumour types.',\n",
       " 'Furthermore, the Pan-Cancer Analysis of Whole Genomes (PCAWG) project has elucidated the extensive heterogeneity and abnormality of whole genomes from cancer cells (Campbell et al., 2020).',\n",
       " 'Thus, NGS continues to transform our ability to study cancer biology.  ',\n",
       " 'Prior to NGS, microarray studies helped to elucidate oncogenic signalling pathways, and the regulatory networks that associate with disease (Curtis et al., 2012; Torrente et al., 2016).',\n",
       " 'Compared to microarray, RNAseq has greater dynamic range, sensitivity and an ability to detect novel transcripts',\n",
       " '37requiring no prior sequence knowledge.',\n",
       " 'RNAseq is capable of quantifying the expression of approximately 20,000 human protein-coding genes and thousands more non-coding transcripts from a single tumour sample allowing the inference of upstream signalling processes and regulatory networks by differential gene expression (DGE) analysis and subsequent gene set enrichment analysis (GSEA) (Subramanian et al., 2005; Robinson, McCarthy and Smyth, 2010; Love, Huber and Anders, 2014).',\n",
       " 'In unsupervised learning, weighted gene co-expression network analysis (WGCNA) supports the identification of transcript modules and so-called ‘hub’ genes or master regulators of transcriptional states from which module-trait relationships can be uncovered (Langfelder and Horvath, 2008).',\n",
       " 'RNAseq is therefore a powerful technique for exploring tumour inflammation.',\n",
       " 'One caveat of this approach is that RNA abundance does not perfectly correlate with protein levels, for example due to degradation or regulation by micro RNAs (miRNAs) (He, Hannon and Harbor, 2004).',\n",
       " 'Moreover, many signalling pathways are controlled by phosphorylation events and other post-translational modifications.',\n",
       " 'Although we can use protein-protein interaction networks to understand the potential impact of gene expression changes on signalling pathways, there is an increasing amount of mass spectrometry-based tumour proteomic and phosphoproteomic data that is widely accessible, for instance via the Clinical Proteomic Tumour Analysis Consortium (CPTAC).',\n",
       " 'Therefore, we can increasingly measure protein signalling cascades directly and on a large scale.',\n",
       " 'Altogether, these resources provide an unprecedented amount of data and potential insight.',\n",
       " 'There are however very few studies that make use of the full extent of publicly available NGS data to search for answers to many of the fundamental questions about how tumours interact with host immune cells.',\n",
       " 'Although several computational methods infer immune cell abundance from bulk tumour transcriptomic data they are inaccurate at detecting lowly abundant intratumoural populations (Jimenez-Sanchez, Cast and Miller, 2019).',\n",
       " 'For example, pan-cancer transcriptome analysis identified γδT cells, that have inadequate markers for IHC (Pont et al., 2012), as being the foremost immune cell population correlated with favourable prognosis (Gentles et al., 2015).',\n",
       " 'However, the CIBERSORT method (Newman et al., 2015) that was used to infer immune cell abundance in the aforementioned study (Gentles et al., 2015) failed to distinguish sorted γδT cells from conventional T cells and NK cells (Tosolini et al., 2017).',\n",
       " 'Indeed, cytotoxic lymphocytes including ILCs, have transcriptomes with high multicollinearity meaning their detection in tumour tissue requires alternative or improved methods.',\n",
       " 'Thus, a major challenge for the field is how to reliably and accurately assess the relationship between rare cell populations or phenotypic states with patient outcomes.',\n",
       " 'scRNAseq largely overcomes these challenges, and many studies have used this approach to uncover the extent of intratumoural cellular heterogeneity in different cancer types compared to healthy tissue, including in the context of ICB (Tirosh et al., 2016; Puram et al., 2017; Jerby-Arnon et al., 2018; Lambrechts et al., 2018; Ma et al., 2019).',\n",
       " 'With this data we can begin to understand not only phenotypic cell states, but also the cellular source of the transcripts identified in bulk tumour analysis to be associated with disease traits.',\n",
       " 'Computational methods have also been developed that elucidate cellular developmental trajectories, future states and cell-cell interactions from single cell transcriptomes (La Manno et al., 2018;',\n",
       " 'Saelens et al., 2019; Browaeys, Saelens and Saeys, 2020).',\n",
       " 'Single cell transcriptomics is therefore helping to unravel how tumour cells progressively skew immune and',\n",
       " '38stromal cell phenotypes to their advantage, and is central towards understanding the interplay of cancer-inhibitory and -promoting inflammation.  ',\n",
       " 'Biomarker assays   Retrospective analysis of clinical trials has shown that patients on biomarker-led trials are more likely to derive clinical benefit (Schwaederle et al., 2016).',\n",
       " 'Approved cancer biomarkers mostly comprise assays such as IHC, in situ hybridisation, serum protein ELISAs and polymerase chain reaction (PCR) reactions to detect mutations.',\n",
       " 'Multi-gene signatures however capture multiple features of a tumour and have great potential in patient risk stratification as demonstrated by the Mammaprint and OncotypeDX assays amongst others (van ’t Veer et al., 2002; Paik et al., 2004; Cardoso et al., 2016; Cieślik and Chinnaiyan, 2018).',\n",
       " 'RNAseq, despite its widespread use in research, has not been successfully leveraged as a clinical diagnostic or prognostic test in cancer.',\n",
       " 'In the clinic, tumour samples are predominantly stored as FFPE tissues.',\n",
       " 'This fixation process leads to nucleic acid degradation over time that impairs our ability to perform NGS experiments.',\n",
       " 'Nanostring gene expression profiling overcomes these issues by using fluorescent barcode-labelled oligonucleotide probes for transcripts of interest making it amenable to highly-degraded nucleic acids and therefore FFPE tissue, and without the need for cDNA conversion that is a feature of PCR based approaches.  ',\n",
       " 'Integrative multi-omics data analysis has fuelled interest in multi-modal biomarkers with examples both in breast ductal carcinoma in situ and in the context of pan-cancer ICB response (Kristensen et al., 2012, 2014; Kaur et al., 2021; Litchfield et al., 2021).',\n",
       " 'However, there remains considerable doubt as to the economic, logistic and experimental feasibility of multi-modal biomarkers meaning there remains interest in the development of quick, simplistic and interpretable predictive models such as those achieved by gene signatures.  ',\n",
       " 'Pan-cancer analysis  Pan-cancer analysis of transcriptional programs is challenging due to profound differences across tissues.',\n",
       " 'Unsupervised analysis of cancer multi-omics data almost perfectly clusters cancers by tissue-of-origin, although a subset of squamous cancers of the lung, bladder, head and neck cluster together indicating commonality among tumour molecular profiles across certain tissues (Hoadley et al., 2014, 2018).',\n",
       " 'Prognostic genes in individual cancer types fail to overlap.',\n",
       " 'However, substantial crossover is evident when genes are compiled into gene sets representing biological pathways and processes (Anaya et al., 2016).',\n",
       " 'Therefore, common principles underpin patient outcomes across cancer types, warranting further investigation of gene signatures for pan-cancer prognostication.',\n",
       " 'Another important consideration in tumour gene expression studies is the landscape of SCNAs (Ciriello et al., 2013; Taylor et al., 2018).',\n",
       " 'It was estimated that 99% of abundantly expressed genes correlate to some extent with copy number, and that approximately 28% of the total variation in gene expression could be explained by gene dosage effects (Fehrmann et al., 2015).',\n",
       " 'As a result, it is essential to examine the genetic factors that underpin transcriptomic measurements.',\n",
       " 'Amidst these challenges, pan-cancer analyses have greatly enhanced our understanding of oncogenic signalling cascades, tumour',\n",
       " '39aneuploidy, cells-of-origin, fusion events, oncogenic viruses, metabolism, enhancer activity, cancer patient prognosis and much more (Campbell et al., 2018; Chen et al., 2018; Gao et al., 2018; Hoadley et al., 2018; Liu et al., 2018; Peng et al., 2018; Sanchez-vega et al., 2018).',\n",
       " 'Multiple studies have used tumour transcriptomes, curated gene sets and unsupervised clustering to identify pan-cancer immune phenotypes with distinct somatic and germline genetics (Thorsson et al., 2018; Shahamatdar et al., 2020; Bagaev et al., 2021; Sayaman et al., 2021).',\n",
       " 'These data-driven approaches provide a detailed overview of the cancer immunogenomic landscape but lack translatability to the clinic.',\n",
       " 'Knowledge-driven pan-cancer analyses, such as those characterising WNT or TGFb pathways (Chakravarthy et al., 2018; Gajewski et al., 2019), have a clearer path towards clinical translation built upon solid, pre-clinical evidence and with a clear strategy of identifying patient populations that may benefit from pathway inhibition.',\n",
       " 'However, no such analysis exists for the COX-2/PGE2 pathway, nor has there been a pan-cancer analysis of the interplay between opposing inflammatory phenotypes and patient outcomes.',\n",
       " 'The primary focus of this PhD thesis is therefore to address these gaps in our understanding by utilising the full extent of bulk tumour and single cell multi-omics data that is available.  ',\n",
       " 'Cyclooxygenase Inflammatory Signature (COX-IS)',\n",
       " 'For the first half of this PhD project, the main objective was to address the question of whether there are qualitatively and quantitatively different types of inflammation, in order to validate our findings using murine cancer models.',\n",
       " 'The results of this analysis were published (Bonavita et al., 2020).',\n",
       " 'Following this publication, we established a set of additional project questions that the main results chapters in this thesis seek to tackle.',\n",
       " 'The findings of the study and the ensuing project questions are now described in detail.',\n",
       " 'COX-deficient (Ptgs-/-) subcutaneous BRAFV600E murine melanomas are a useful tool for studying key determinants of tumour immunogenicity in a setting where tumours have identical TMB, where tumour fate is inextricably linked to its inflammatory composition (Zelenay et al., 2015).',\n",
       " 'These tumours are rejected by immune-competent hosts in a fashion dependent on adaptive immunity, and on cross-presenting, Batf3-dependent DCs (Zelenay et al., 2015).',\n",
       " 'Primarily this work established the COX pathway as a dominant regulator of immune evasion and a therapeutic target for enhancing the efficacy of PD1 blockade (Zelenay et al., 2015).',\n",
       " 'COX-1 and COX-2 isoforms convert arachidonic acid into prostaglandin-H2 (PGH2), which is converted into PGE2 by the prostaglandin E synthase enzymes, encoded by PTGES1-3 genes.',\n",
       " 'Whereas COX-1 is constitutively expressed, COX-2 is an inducible isoform believed to be the rate-limiting step in production of PGE2 (Greenhough et al., 2009; Kalinski, 2013).',\n",
       " 'Aspirin is a non-selective inhibitor of COX enzymes, whilst celecoxib targets COX-2 specifically.',\n",
       " 'Although COX enzymes promote the production of other prostanoids, PGE2 is considered as the foremost of these by importance in cancer (Wang and Dubois, 2010).',\n",
       " 'Although, the contribution of immune evasion towards the pro-tumourigenic activity of COX-2/PGE2 was elucidated (Zelenay et al., 2015), the exact mechanism by which PGE2 operates intratumourally to subvert anti-tumour immunity remained elusive.  ',\n",
       " 'To unravel this mechanism, as well as define the molecular and cellular features of tumours with divergent fate, we studied wild-type (WT) and COX-deficient BRAFV600E melanomas at an early',\n",
       " '40time point before the onset of adaptive immunity (Zelenay et al., 2015).',\n",
       " 'Reproducibly across multiple murine models, NK cells were more abundant in COX-deficient compared to WT tumours (Bonavita et al., 2020).',\n",
       " 'This phenotype was reversed by genetically restoring COX activity in deficient cancer cells.',\n",
       " 'Depletion of NK cells at day 4, but not at day 7, post-implantation led to a clear increase in the size of COX-deficient tumours.',\n",
       " 'Therefore, there was a striking inverse relationship between COX activity and early NK cell infiltration of tumours which dictated tumour fate.',\n",
       " 'Although, NK cells were involved in innate control of tumours early on in development through direct killing of COX-deficient tumour cells, there was a more pronounced role for NK cells in later, T cell-mediated adaptive control of tumours (Bonavita et al., 2020).',\n",
       " 'By using a genetically engineered mouse model (GEMM) in which NK cells were rendered insensitive to PGE2 (Gzmb-Cre Ptger4floxed/floxed Ptger2-/- or GPP mice for short) we showed that tumour progression was dependent on cancer cell-derived PGE2 directly suppressing NK cell cytotoxicity and IFNg secretion.',\n",
       " 'We used bulk tumour and scRNAseq to show that NK cell activity predominantly via IFNg signalling skewed intratumoural myeloid populations towards a more favourable phenotype preceding the onset of adaptive immunity (Bonavita et al., 2020).',\n",
       " 'From NK cell depletion experiments we derived an inflammatory transcriptional signature of genes dependent on early NK cell reprogramming of the TME.',\n",
       " 'This gene signature was defined as the Cancer Inhibitory (CI) inflammation signature.',\n",
       " 'Likewise, inflammatory genes that depended on cancer cell-intrinsic COX/PGE2 activity in the TME were also identified and the resulting gene set was defined as the Cancer Promoting (CP) inflammation signature.',\n",
       " 'Thus, we uncovered signatures of antagonistic flavours of inflammation.  ',\n",
       " 'The first aim of this PhD project was to validate these findings in humans.',\n",
       " 'Consistent with our murine models, COX-2 mRNA expression (PTGS2) was highly correlated with genes of the CP signature across all TCGA datasets (Bonavita et al., 2020).',\n",
       " 'There was also a powerful anti-correlation of the CI signature genes with PTGS2 expression in several cancer types.',\n",
       " 'In agreement with our model, NK cells were highly correlated to CI signature genes.',\n",
       " 'Therefore, human tumours with elevated levels of COX-2 mRNA were devoid of transcripts that were downstream of NK cell-mediated TME reprogramming in our murine models.  ',\n",
       " 'We next hypothesised that the balance of CP and CI signatures would identify patients with distinct prognosis consistent with the opposite fates of WT and COX-deficient tumours.',\n",
       " 'For this purpose, we computed a ratio of the CP and CI signatures, referred to as the Cyclooxygenase Inflammatory Signature (COX-IS) and performed patient survival analysis.',\n",
       " 'Compared to individual signature genes, the CP and CI signatures alone, and signatures of IFNg signalling, NK cells and T cells, the COX-IS was more consistently and powerfully prognostic (Bonavita et al., 2020).',\n",
       " 'In multivariate survival analysis, COX-IS had independent prognostic value adjusted for age, sex, stage, and cancer type-specific characteristics (Bonavita et al., 2020).',\n",
       " 'Moreover, across nine independent, transcriptomic cohorts of patients treated with ICB, the COX-IS was consistently correlated with poor response to ICB (Bonavita et al., 2020).',\n",
       " 'Multivariate logistic regression in metastatic urothelial carcinoma (mUC) and ccRCC datasets confirmed that COX-IS explained a significant proportion of the variance in patient response that could not be explained by typical immune biomarkers such as TMB and PD-L1 IHC.',\n",
       " 'Moreover, COX-IS high melanoma and mUC patients had significantly worse OS',\n",
       " '41following ICB than COX-IS low patients (Bonavita et al., 2020).',\n",
       " 'Collectively, these findings suggested that the balance of antagonistic inflammatory phenotypes is a critical determinant of patient survival and benefit from ICB, which supported a patent filing for the COX-IS (WO2019243567A1).  ',\n",
       " 'By studying the mechanistic basis for immune evasion dictated by activation of the COX-2/PGE2 pathway we found a critical role for NK cells early in tumour development.',\n",
       " 'In the absence of PGE2 signalling via EP2/4, these cells promoted an inflammatory landscape that is geared towards endogenous and therapy-induced anti-tumour immunity.',\n",
       " 'This mechanistic link between COX-2/PGE2 and EP receptor signalling in NK cells was based upon experiments using GPP mice.',\n",
       " 'These mice lack germline EP2 receptor expression, and EP4 on GZMB+ cells that were almost exclusively NK cells in our tumour model settings.',\n",
       " 'With this model, we could not rule out a pro-tumourigenic role for PGE2 signalling via EP2 directly on other host cell types.',\n",
       " 'Therefore, it remains important to uncover whether PGE2 signalling through EP2 on DCs, macrophages, CAFs and cancer cells also contributes to immune evasion.                          ',\n",
       " '42Project hypotheses and objectives Our study (Bonavita et al., 2020) did not formally address and quantify the individual contribution of cancer-promoting inflammation towards the prognostic and predictive power of the COX-IS nor whether the denominator of the COX-IS (CI signature) was interchangeable with other biomarkers of ‘hot’  tumours such as the TIS signature or TMB (Ayers et al., 2017; Cristescu et al., 2018).',\n",
       " 'Moreover, our benchmarking did not extend to other resistance-associated signatures such as stroma-EMT (Wang et al., 2019) or TGFβ signalling (Chakravarthy et al., 2018; Mariathasan et al., 2018).',\n",
       " 'In addition, we did not address the genetic features of tumours with high levels of cancer-promoting inflammation.',\n",
       " 'There is a flawed assumption in the field that ‘inflamed’ tumours generally respond better to ICB than tumours devoid of immune cells.',\n",
       " 'We sought to challenge this dogma by analysing the interplay of cancer-promoting inflammation with the type of inflammation that is widely considered as cancer-restrictive, and associated with extended patient survival and ICB sensitivity.',\n",
       " 'As a tool for studying COX-2/PGE2-associated inflammation, the CP signature (Bonavita et al., 2020) may be limited in its relevance and prognostic power by its murine derivation.',\n",
       " 'As such, we aimed first to derive a new, human signature of COX-2-centred pro-tumourigenic inflammation.',\n",
       " 'Subsequently, we developed the following set of core project questions designed to comprehensively and systematically evaluate the nature and role of pro-tumourigenic inflammation in cancer progression and response to ICB (Figure 0.5).  ',\n",
       " '(1) Is COX-2/PGE2-associated inflammation pro-tumourigenic and correlated with disease progression and poor prognosis?',\n",
       " '(2) What are the molecular characteristics of tumours with antagonistic inflammatory phenotypes?',\n",
       " '(3) What is the cellular composition of tumours with high levels of cancer-promoting or -inhibitory inflammation?',\n",
       " '(4) What are the oncogenic events and signalling networks that promote cancer cell-intrinsic expression of pro-tumourigenic inflammatory mediators?',\n",
       " '(5) Does pro-tumourigenic inflammation suppress anti-tumour immunity in humans and what is its effect on the spatial and molecular features of tumour-infiltrating lymphocytes?',\n",
       " '(6) Does measurement of pro-tumourigenic inflammation add predictive value to current biomarkers of response to checkpoint inhibitors? ',\n",
       " 'To address these questions, we used publicly available datasets, in vitro experiments and retrospective analysis of cancer patient samples from the Manchester Cancer Research Centre human tissue biobank.',\n",
       " 'The main results from these investigations will be described in the proceeding four chapters of this thesis.     ',\n",
       " '43 Figure 0.5: Project questions.',\n",
       " 'A figure illustrating a tumour microenvironment, on the left with pro-tumourigenic features, and on the right with cancer-restrictive features.',\n",
       " 'Project questions are depicted in brackets and by question marks.                     ',\n",
       " '44                 Chapter 1 Pro-tumourigenic inflammation suppresses anti-tumour immunity and promotes poor prognosis          ',\n",
       " '45ABSTRACT  On the one hand, tumour-associated inflammation is a hallmark of aggressive and therapy-resistant tumours, and on the other, inflammatory responses mediated by select immune cells drive tumour control and associate with favourable prognosis.',\n",
       " 'Here, we investigated the interplay between pro-tumourigenic inflammation (PTI) and T cell-mediated immunity through analysis of multiple cancer patient datasets.',\n",
       " 'We computationally defined a COX-2/PGE2 pathway centred, PTI gene signature that shows profound prognostic utility across pan-cancer immune subtypes and individual malignancies after adjusting for clinical parameters such as age, sex and disease stage.',\n",
       " 'Individual elements of the signature were prognostic across different cancer tissues, but less consistently than PTI itself.',\n",
       " 'Moreover, we uncover an intimate interaction of PTI with IFNg signalling that highlights a major contribution for immune evasion in the tumour-promoting role of inflammation.',\n",
       " 'Accordingly, PTI identifies patients with dismal outcome despite harbouring ‘hot’ tumours based on established immune biomarkers such as tumour-infiltrating lymphocytes, CD8+ T cells, PD-L1, IFNg signalling and tumour mutational burden.',\n",
       " 'Our findings therefore elucidate the dual role of inflammation in cancer and identify PTI as an independent biomarker to predict cancer patient survival.      ',\n",
       " '46INTRODUCTION  A common feature of tumours responsive to ICB, and associated with long-term survival of cancer patients, is T cell inflammation that is a surrogate of so-called ‘hot’ tumours (Fridman et al., 2012; Tumeh et al., 2014; Ayers et al., 2017; Danaher et al., 2018).',\n",
       " 'Yet, there are abundant examples of highly CD8+ T cell infiltrated tumours progressing rapidly, and conversely of ‘cold’ tumours responding vigorously to ICB.  ',\n",
       " 'Inflammation is arguably a feature of most clinically apparent tumours, except the type of intratumoural inflammatory response more commonly found fuels cancer growth and contributes to resistance to immunotherapy (Mantovani et al., 2008; Coussens, Zitvogel and Palucka, 2013; Shalapour and Karin, 2019).',\n",
       " 'In recent work, we found that quantitatively and qualitatively different inflammatory profiles depended on the presence or absence of cancer cell-intrinsic COX-2 expression (Zelenay et al., 2015; Bonavita et al., 2020) and that pharmacological inhibition of the COX-2 pathway skews the TME towards cancer-restrictive inflammation.',\n",
       " 'By interrogating large patient datasets, we demonstrated how a murine-derived gene signature combining measurement of opposing inflammatory flavours had striking prognostic and predictive power in retrospective analysis including following ICB.',\n",
       " 'As such, the level of activity in the COX-2/PGE2 pathway, upregulated in numerous cancers and implicated in various aspects of malignant growth constitutes a candidate for molecular stratification of the inflammatory phenotype of tumours (Dannenberg and Subbaramaiah, 2003; Wang and Dubois, 2010; Zelenay et al., 2015).',\n",
       " 'Our previous bioinformatics analysis focused on combining cancer-promoting and -inhibitory inflammatory signatures into a single score; termed the Cyclooxygenase Inflammatory Signature (COX-IS) (Bonavita et al., 2020).',\n",
       " 'Thus far, there has been no deep, in silico characterization of COX-2/PGE2-centred inflammation as an independent cancer hallmark and prognostic factor across diverse contexture.',\n",
       " 'Herein, we constructed a bioinformatics workflow to investigate the prognostic and predictive power of COX-2-centred pro-tumourigenic inflammation in human tumours.',\n",
       " 'Utilising pan-cancer transcriptomic data, we identified extracellular signalling molecules common to cancers with COX-2 expression but not normal tissues.',\n",
       " 'We used this signature as a tool to investigate the prognostic power of pro-tumourigenic inflammation in the context of long-term patient survival across multiple cancer types and immune subtypes of cancer.',\n",
       " 'Meanwhile, we investigate the interaction of tumour-promoting inflammation with biomarkers of so-called ‘hot’ or immunogenic tumours such as gene signatures of T cell inflammation and IFNg signalling.       ',\n",
       " '47RESULTS  A signature of extracellular ligands linked to COX-2 expression captures pro-tumourigenic inflammation  Based on the reported high prevalence of COX-2/PGE2 activity in cancer and our work establishing the pathway as a nodal regulator of immune evasive inflammation, we hypothesised that transcripts associated with high COX-2 mRNA (PTGS2) levels would capture conserved, pan-cancer features of a pro-tumourigenic inflammatory signalling network (Dannenberg and Subbaramaiah, 2003; Wang and Dubois, 2010).',\n",
       " 'As transcript levels of PTGS2 were heterogeneous between and within cancer types (Figure 1.1A), we devised a computational pipeline that compared PTGS2 high and low tumours from the top ten PTGS2-expressing cancer types (Figure 1.1A, B).',\n",
       " 'Among the 54 genes overexpressed in COX-2high cancers (Figure 1.1C) were classic secreted factors implicated in local and systemic inflammation such as IL6, CXCL8, IL1B and CXCL2, cell surface receptors (CSF3R, IL7R, CD69), neutrophil markers (G0S2) as well as epidermal growth factor receptor (EGFR) ligands and growth factors such as AREG and EREG, in addition to ECM modifiers HAS2 and MMP12.',\n",
       " 'GSEA using Hallmark pathway signatures revealed that the TNF-NFκB pathway and other inflammatory pathways (IL6-STAT3) were upregulated in COX-2high cancers (Figure 1.1D).',\n",
       " 'Interestingly, EMT, complement, hypoxia, KRAS signalling and apoptosis pathways were also overexpressed (FDR< 0.05, normalised enrichment score (NES) > 1).',\n",
       " 'MYC targets, adipogenesis and oxidative phosphorylation pathways were enriched in COX-2low tumours (FDR< 0.05, NES< -1).',\n",
       " 'This data therefore strengthens the notion that the COX-2 pathway, using PTGS2 transcript levels as a surrogate, represents a central node within inflamed tumours, and that COX-2high tumours are enriched in various biological processes linked to cancer hallmarks.            ',\n",
       " '48   Figure 1.1: Derivation of a pan-cancer pro-tumourigenic inflammatory signature linked to COX-2.',\n",
       " '(A) COX-2 mRNA',\n",
       " '(PTGS2) expression across 31 TCGA datasets.',\n",
       " '(ccRCC = clear cell renal cell carcinoma, ACC = adrenocortical carcinoma, UCS = uterine carcinosarcoma, ESCA = oesophageal cancer, UCEC = uterine and endometrial cancer, LAML = acute myeloid leukaemia, LGG = lower grade glioma, LUSC = lung squamous cell carcinoma, SKCM = skin cutaneous melanoma, CHOL = cholangiocarcinoma, PAAD = pancreatic adenocarcinoma, MESO = mesothelioma, SARC = sarcoma, COAD = colon adenocarcinoma, READ = rectal adenocarcinoma, THCA = thyroid cancer, KIRP = papillary renal cancer, GBM = glioblastoma, LIHC = liver hepatocellular carcinoma, PCPG = pheochromocytoma, PRAD = prostate cancer, OV = ovarian cancer, BLCA = bladder cancer, BRCA = breast cancer, KICH = chromophobe renal cancer, UVM = uveal melanoma, THYM = thymoma, TGCT = testicular cancer, DLBC = diffuse large B cell lymphoma, STAD = stomach adenocarcinoma, CESC = cervical cancer, HNSC = head and neck cancer) (B) Bioinformatic pipeline used to discover a gene signature of COX-2-centred PTI.',\n",
       " '(C) Volcano plot of differentially-expressed genes (DEGs) comparing PTGS2 high and low tumours across ten cancer types.',\n",
       " 'Significantly upregulated genes are marked in red and downregulated in blue using 1.2 < log2Fold Change (FC) <',\n",
       " '-1.2, FDR< 0.05 cut-off values.',\n",
       " '(D) Normalised enrichment scores from GSEA using DEGs from comparing PTGS2 high and low cancers.  ',\n",
       " 'To further address whether the genes upregulated in COX-2high tumours were members of a biologically-meaningful and cohesive signalling cascade, we constructed a protein-protein interaction  ',\n",
       " '49network using the curated STRING database (Szklarczyk et al., 2019).',\n",
       " 'Consistent with their shared involvement in signalling pathways, proteins encoded by genes upregulated in COX-2high tumours had significantly more interactions than would be expected by chance (P< 1E-16) with an average of 9.7 edges per node.',\n",
       " 'These proteins were overrepresented in gene ontology biological processes such as the inflammatory response, myeloid cell chemotaxis, response to cell death and stress (Figure 1.2B).',\n",
       " 'Next, we questioned whether the COX-2 protein interaction network was qualitatively different in tumours compared to healthy tissue.',\n",
       " 'As such, we performed DGE analysis between COX-2high and COX-2low tumour-adjacent healthy tissues as we did for tumour tissue.',\n",
       " 'We constructed a protein-protein interaction network that, similarly to the tumour network, had a significant enrichment of interactions (P< 1E-16).',\n",
       " 'In contrast to the tumour tissue network, the normal tissue COX-2 interaction network was enriched for pathways related to tissue homeostasis and cellular differentiation (Figure 1.2D).',\n",
       " 'However, there were significantly fewer interactions than in the tumour network at an average of 3.9 per node (P= 3.4E-7) (Figure 1.2D).',\n",
       " 'As such, proteins in the tumour network interact more closely with one another suggesting that they may operate in a more coordinated fashion.',\n",
       " 'Statistical analysis of each network revealed how COX-2 was more central in the tumour network with higher neighbourhood connectivity (mean nearest neighbour number of edges) and hub score (Figure 1.2E).',\n",
       " 'These findings suggest that COX-2 may have greater influence over the tumour network.',\n",
       " 'Altogether, these results highlight how tumours are rewired with COX-2 activity at the epicentre of a highly coordinated signalling network that responds to high levels of cellular stress and death by promoting inflammation.',\n",
       " 'On the other hand, in normal tissues, COX-2 is uncoupled from inflammatory processes and linked instead to homeostatic signalling.',\n",
       " 'Thus, the COX-2 network is qualitatively different in tumours relative to non-tumour tissue.                ',\n",
       " '50      ',\n",
       " '51Figure 1.2: Tumour and normal tissue COX-2high protein-protein interactions network analysis.',\n",
       " '(A) Protein-protein interaction network for genes upregulated in COX-2high tumours pan-cancer, or (C) tumour-adjacent normal tissues.',\n",
       " '(B) Significantly enriched gene ontology biological processes in the tumour and (D) normal tissue networks.',\n",
       " '(E) Comparison of the number of edges (otherwise known as predicted interactions) per node (protein) in the normal and tumour tissue networks.',\n",
       " 'P value refers to a Wilcoxon test.',\n",
       " '(G) Table outlining metrics related to COX-2 in the normal and tumour networks including barplot showing Kleinberg’s hub centrality scores for all network genes along the x axis.',\n",
       " 'PTI genes are shown in red.  ',\n",
       " 'A central aim of this project was to elucidate the instructive signals of a pro-tumourigenic inflammatory signalling network.',\n",
       " 'We postulated that genes encoding soluble factors capable of delivering autocrine, paracrine and endocrine signals that re-model the local and systemic inflammatory composition would be strong candidates as representative nodes of a hostile inflammatory TME.',\n",
       " 'As such, we filtered the upregulated genes by a list of predicted ligands (Ramilowski et al., 2015).',\n",
       " 'The resulting 16 genes included chemokines (CXCL1, CXCL2, CXCL3, CXCL5, CXCL8), cytokines and growth factors (IL1A, IL1B, IL6, NRG1) many of which reassuringly have previous links to cancer and inflammation (Figure 1.3), as well as being highly connected nodes in the COX-2high tumour protein interaction network (Figure 1.2A).',\n",
       " 'Henceforth this gene set will be referred to as the Pro-Tumourigenic Inflammation (PTI) signature.      ',\n",
       " 'Figure 1.3: COX-2-associated ligand-receptor interactions.',\n",
       " 'An illustration of the PTI signature genes, their receptor interactions and reported roles in cancer.',\n",
       " 'AA = arachidonic acid.',\n",
       " 'PTGES = prostaglandin E synthase enzyme.',\n",
       " 'TXA2 = thromboxane-A2.',\n",
       " 'COX-2/PGE2 (Xu et al., 2014; Zelenay et al., 2015; Böttcher et al., 2018; Zheng et al., 2019; Bonavita et al., 2020; Roulis et al., 2020; Gonçalves et al., 2021; Pelly et al., 2021).',\n",
       " 'CXCL1 (Li et al., 2018; Nishida et al., 2020; Gonçalves et al., 2021).',\n",
       " 'CXCL2/6',\n",
       " '(Liao et al., 2019).',\n",
       " 'CXCL5 (Hsu et al., 2013; Chen et al., 2019; Li et al., 2019).',\n",
       " 'IL-1a and IL-1b (Kumar et al., 2012; Chittezhath et al., 2014; Voigt et',\n",
       " 'al., 2017; Wu et al., 2018; Biffi et al., 2019).',\n",
       " 'IL-6',\n",
       " '(Che et al., 2017; Hayashi et al., 2018; Tsukamoto et al., 2018; Del Poggetto et al., 2021).',\n",
       " 'IL-8',\n",
       " '(Sparmann and Bar-Sagi, 2004; Nishida et al., 2020; Schalper et al., 2020; Yuen et al., 2020).',\n",
       " 'IL-11 (Calon et al., 2012; Huynh et al., 2021; Wang et al., 2021).',\n",
       " 'NRG1 (Cabrera et al., 2018; Zhang et al., 2020).',\n",
       " 'HAS2 (Okuda et al., 2012; Elyada et al., 2019).',\n",
       " 'AREG (Green et al., 2017; Shimada et al., 2017;  ',\n",
       " '52Xu et al., 2019).',\n",
       " 'EREG (He et al., 2019; Wang et al., 2019).  ',\n",
       " 'MMP12 (Nabha et al., 2008; Nocito et al., 2008).',\n",
       " 'VCAN (Kim et al., 2009; Hope et al., 2017; Zhang et al., 2019).  ',\n",
       " 'Two of the most abundant, non-malignant cell types occupying the TME are fibroblasts and macrophages.',\n",
       " 'To investigate whether COX-2-driven PGE2 activity on fibroblasts and macrophages might explain the association of individual PTI elements with COX-2 expression in tumours, we treated murine BMDMs in vitro with synthetic PGE2.',\n",
       " 'In addition, we separately analysed published datasets of PGE2-treated human fibroblasts and CD4+ T cells (Wettlaufer et al., 2016).',\n",
       " 'Notably, the majority of ligands were significantly induced by PGE2 either in fibroblasts or macrophages (Figure 1.4).',\n",
       " 'A number of ligands were also induced by PGE2 treatment of human CD4+ T cells (Figure 1.4).',\n",
       " 'These findings validate our bioinformatic pipeline and suggest that the PTI signature of ligands represents nodes hierarchically downstream of the COX-2/PGE2 axis that functions as a ‘hub’ in the network.    ',\n",
       " 'Figure 1.4:',\n",
       " 'PGE2 upregulates transcription of pro-tumourigenic inflammatory ligands in multiple cell types.',\n",
       " 'Log2FC of PTI genes from DGE analysis of mouse BMDMs, human fibroblasts and CD4+ T cells treated in vitro with synthetic PGE2.',\n",
       " 'Significantly upregulated genes (FDR< 0.05, log2FC> 1) are in red and downregulated genes are in blue (FDR< 0.05, log2FC< -1).  ',\n",
       " 'IFNg signalling is antagonised by concomitant PTI expression  Next, given the enrichment of cancer hallmarks in COX-2high tumours, we examined the association between PTI levels and patient overall survival (OS); that is typically the time between diagnosis and patient death.',\n",
       " 'Using TCGA data, we performed pan-cancer Cox Proportional Hazards (CoxPH) regression analysis correcting for disease stage and cancer type.',\n",
       " 'CoxPH models allow us to estimate the effect of multiple covariates, such as patient age at diagnosis and tumour stage, on the likelihood of a particular event (for example death) happening at a specific time.',\n",
       " 'The hazard ratios (HR) of covariates in a CoxPH model correspond to exponentiated coefficients or effect sizes.',\n",
       " 'For example,  ',\n",
       " '53if the HR for PTI as a continuous variable is 2 then, if all other covariates are kept constant, there is a two-fold increase in the hazard for a one value increase in PTI level at any point in time.',\n",
       " 'Indeed, we found that the PTI signature correlated strongly with shortened survival of cancer patients both as a continuous variable (HR= 1.23, P= 1E-05) and stratifying according to quartiles (PTIhigh versus PTIlow, HR= 1.25, P= 0.02) (Figure 1.5A).',\n",
       " 'As such, we concluded that the transcriptional information encoded by PTI ligands must constitute a conserved poor prognosis tumour phenotype.',\n",
       " 'Next, we sought to study the interaction of PTI with signatures of cancer-inhibitory inflammation.',\n",
       " 'To test whether the strength of association between PTI and poor prognosis (Figure 1.5A) differs by the level of cancer-inhibitory inflammation, we analysed the interaction term of the PTI signature, with T cell inflammation (TIS), IFNg signalling and CD8+ T cell signatures in multivariate CoxPH models predicting patient OS (Ayers et al., 2017; Danaher et al., 2017).',\n",
       " 'By including linear multiplicative interaction terms (for example the product of PTI and TIS signatures) in our models, we can estimate whether the effect of one covariate on survival depends on values of the other covariate.',\n",
       " 'Strikingly, each signature interacted significantly with PTI (Figure 1.5B, Supplementary Table 1.1, 1.2).',\n",
       " 'As anticipated, PTI generally associated with increased mortality (HR= 1.25, P< 0.0001).',\n",
       " 'Therefore, for a given patient with mean levels of IFNg signalling, a one value increase in PTI correlates with a 25% increase in risk.',\n",
       " 'By contrast, for a patient with mean levels of PTI, an incremental increase in IFNg signalling corresponds to a 16% decrease in the risk of death (HR=0.84, P<0.0001, Figure 1.5B).',\n",
       " 'Meanwhile, there was a significant interaction term for PTI and IFNg signatures (HR= 1.13, P= 0.004).',\n",
       " 'To illustrate this interaction, we simulated the hazard at high and low expression levels of IFNg signalling (Figure 1.5C).',\n",
       " 'At low levels of IFNg signalling the hazard rate was consistently elevated regardless of PTI level, whereas for patients with high IFNg signalling the risk of death increased strongly with PTI expression (Figure 1.5C).',\n",
       " 'Indeed, at a high level of PTI (for example PTI score= 1.5, Figure 1.5C) a patient with extensive evidence of IFNg signalling in the TME, that is typically a marker of better prognosis tumours, would have a similar risk of death to a patient with a non-T cell-inflamed tumour.',\n",
       " 'Thus, high intratumoural IFNg signalling is only good news for a patient if PTI levels are simultaneously low.',\n",
       " 'Similar interaction effects were observed between PTI and CD8+ T cell and TIS signatures (Supplementary Table 1.1, 1.2).',\n",
       " 'Collectively, these findings illuminate the exquisite interaction between pro- and anti-tumourigenic inflammatory phenotypes highlighting the profound antagonistic effect of the PTI signature towards T cell-inflamed tumours.    ',\n",
       " '54Figure 1.5: Pro-tumourigenic inflammation interacts with IFNg signalling and promotes poor prognosis.',\n",
       " '(A) Kaplan-Meier (KM) plot of pan-solid tumour TCGA patients stratified into quantiles by PTI signature score with associated risk table.',\n",
       " '(B) HRs and 95% confidence intervals for continuous PTI and IFNg signatures, as well as their interaction term, from a multivariate CoxPH model with cancer type and disease stage strata pan-cancer.',\n",
       " '(C)',\n",
       " 'A simulation of the hazard rate for exemplar high and low IFNg patients across a range of PTI values.',\n",
       " 'Lines indicate median hazard rates.',\n",
       " 'Darker shaded areas are quartiles and lighter shaded areas indicate minimum and maximum hazard rates.    ',\n",
       " 'The TME and immune infiltrate vary greatly across tumour types, yet pan-cancer immunogenomic analysis has identified six distinct immune subtypes namely ‘Wound healing’, ‘IFNγ-dominant’, ‘Inflammatory’, ‘Lymphocyte-depleted’, ‘Immuno-quiet’, and ‘TGFβ-dominant’ groups (Thorsson et al., 2018).',\n",
       " 'These subtypes span multiple cancer tissue types and molecular subtypes and were found to have unique genomic, transcriptomic and clinical characteristics (Thorsson et al., 2018).',\n",
       " 'We questioned whether PTI or signatures of ‘hot’ tumours were surrogates of these immune subtypes (Figure 1.6A).',\n",
       " '‘Wound healing’, ‘IFNγ-dominant’ and ‘TGFβ-dominant’ groups had comparably high PTI scores illustrating that equivalent expression levels for pro-tumourigenic inflammatory mediators can be achieved across cancers with diverse background and contexture (Figure 1.6A).',\n",
       " 'Signatures of ‘hot’ tumours were significantly overexpressed in ‘IFNγ-dominant’ subtype tumours highlighting how this pan-cancer immune subtype represents the group of patients deemed most likely to benefit from ICB according to the current paradigm (Figure 1.6A).  ',\n",
       " 'Next, we examined the prognostic utility of PTI across these diverse immune subtypes.  ',\n",
       " 'Crucially, PTI was associated with worse survival in all immune subtypes, not reaching statistical significance only for the ‘Immuno-quiet’ subtype for which PTI levels were lowest (Figure 1.6B).',\n",
       " 'In contrast, biomarkers used to identify ‘hot’ tumours like TIS, CD8+ T cell, and IFNg signalling signatures, or biomarkers of tumour antigenicity such as non-silent mutation rate, referred to here as TMB, and single nucleotide variant neoantigen burden (SNV NAB) were infrequently prognostic across the subtypes (Figure 1.6B)',\n",
       " '(Thorsson et al., 2018).',\n",
       " 'Moreover, these supposed indicators of more immunogenic tumours were inconsistently associated with survival (Figure 1.6B).',\n",
       " 'In ‘Inflammatory’ and ‘Lymphocyte-depleted’ tumours, for example, the TIS correlated with poor survival outcomes (Figure 1.6B).',\n",
       " 'These findings challenge the misconception in the field that the abundance of neoantigens, the presence of infiltrating T lymphocytes, and their activation determines prognosis and long-lasting tumour immunity, and rather highlight the profound, consistent impact of PTI on patient outcomes.',\n",
       " 'Indeed, PTI can be a prognostic indicator across tumour-immune subtypes with altogether very contrasting features, whereas typical markers of so-called ‘hot’ tumours such as IFNg signalling predominantly correlated with prolonged OS in a single immune subtype.  ',\n",
       " 'Once more, we used multivariate survival analysis to test whether PTI and IFNg signalling interact across different immune subtypes.',\n",
       " 'This revealed subtype-specific phenomena underpinned by a commonality in which PTI was consistently and powerfully prognostic even amongst patients with high levels of IFNg signalling (Supplementary Figure 1.1A).',\n",
       " 'In ‘Wound healing’ tumours there was a comparable interaction of PTI and IFNg as in the pan-cancer analysis (Supplementary Figure 1.1A, Supplementary Table 1.3).',\n",
       " 'Likewise, in ‘Lymphocyte-depleted’ tumours there was also a significant',\n",
       " '55interaction, albeit this time PTI correlated with poor prognosis predominantly at low levels of IFNg signalling (Supplementary Figure 1.1A).',\n",
       " 'This may reflect the background, negative prognostic effect of PTI that is independent of its role in promoting immune evasion.',\n",
       " 'In ‘IFNγ-dominant’ patients for whom CD8+ T cells demonstrate unequivocal association with better outcome, as is often the case for ICB-treated patients, PTI was associated with poor survival at either high or low IFNg signalling levels (Figure 1.6C).',\n",
       " 'Indeed, PTIhighIFNghigh patients had a median OS of 5 years compared with >15 years in the PTIlowIFNghigh group despite equivalently high levels of IFNg and CD8+ T cell infiltration (Figure 1.6D).',\n",
       " 'These results suggest that the benefits of IFNg signalling are negated by concomitant PTI expression.',\n",
       " 'This finding was observed across multiple immune subtypes (Supplementary Figure 1.1A), and when substituting the IFNg signalling signature with TIS, CD8+ T cell, PD-L1, as well as tumour infiltrating lymphocyte (TIL) fraction (Supplementary Figure 1.1B) (Ayers et al., 2017; Danaher et al., 2017; Thorsson et al., 2018).',\n",
       " 'Therefore, COX-2-centred PTI antagonises the beneficial effect of immune biomarkers associated with ‘hot’ T cell-inflamed tumours.     ',\n",
       " 'Figure 1.6: Pro-tumourigenic inflammation consistently associates with poor prognosis across diverse pan-cancer immune subtypes.',\n",
       " '(A) Boxplots showing scores for PTI, CD8+ T cell, T cell inflammation (TIS) and IFNg signalling signatures across Thorsson et al (2018) pan-cancer immune subtypes.',\n",
       " 'Non-parametric T test results are shown.',\n",
       " '(B) HRs and 95% confidence intervals from univariate Cox regression survival analysis for biomarkers of interest in six immune subtypes.',\n",
       " '(C) KM survival curve (left panel) for ‘IFNg- dominant’ subtype patients stratified into four groups by PTI and IFNg signatures.',\n",
       " 'Asterisks refer to FDR values.',\n",
       " 'Boxplots (right panel) comparing CD8+ T cell scores between groups.',\n",
       " 'All P values are represented by asterisks unless otherwise stated (*<0.05, **<0.01, ***<0.001).  ',\n",
       " '56Pro-tumourigenic inflammatory mediators are powerful indicators of poor prognosis  To account for potential confounding effects of organ-specific immune features (Salmon et al., 2019), we examined the interaction between PTI and IFNg signalling in individual cancer types using multivariate CoxPH models with and without the linear multiplicative interaction term.',\n",
       " 'In nine cancer types, we found that the PTI signature either as a single variable or through its interaction with IFNg signalling was a statistically significant prognostic factor when adjusted for clinical factors such as age, sex and disease stage (Figure 1.7).',\n",
       " 'Increasing PTI was associated with higher mortality regardless of the level of IFNg signalling in pancreatic cancer (PAAD), ccRCC, lung adenocarcinoma (LUAD), liver cancer (LIHC) and cervical cancer (CESC) (Figure 1.7).',\n",
       " 'In thyroid cancer (THCA), head and neck cancer (HNSC), and uterine cancer (UCEC) there was a significant interaction of PTI and IFNg signalling, thereby mimicking observations in the pan-cancer dataset (Figure 1.7).',\n",
       " 'Non-significant trends were notable in multiple other cancer types (not shown).',\n",
       " 'As such, we concluded that the prognostic power of PTI extends to individual malignancies, and crucially is frequently linked to its suppressive effect on IFNg signalling.  ',\n",
       " 'Figure 1.7: PTI has independent prognostic power in individual cancer types.',\n",
       " 'HRs, 95% confidence intervals and Wald test P values from multivariate CoxPH models of OS in ccRCC (n= 531), LIHC (n= 374), PAAD (n= 176), glioblastoma (GBM, n= 168), CESC (n= 306), adrenocortical carcinoma (ACC, n= 79), UCEC (n= 277), THCA (n= 513) and HNSC (n= 240).   ',\n",
       " '57By performing survival analysis for individual PTI genes, we next assessed whether there was a dominant factor explaining the prognostic power of the PTI signature.',\n",
       " 'The COX-2 gene (PTGS2) itself indicated poor survival in ccRCC patients (Figure 1.8).',\n",
       " 'EGFR ligands EREG and AREG were associated with poor survival in cancers with prevalent KRAS mutations (PAAD, LUAD).',\n",
       " 'Cytokines IL11 and IL1A were prognostic in LGG and ccRCC, whilst chemokines CXCL5 and CXCL1 were prognostic in CESC and ccRCC respectively (Figure 1.7).',\n",
       " 'This data highlights how individual inflammatory ligands heterogeneously contribute towards the association of PTI with poor survival across multiple malignancies, and no single gene is as consistently prognostic as the signature itself.   ',\n",
       " 'Figure 1.8: Individual PTI genes are heterogeneously prognostic across cancer types.',\n",
       " 'HRs and 95% confidence intervals from univariate Cox regression analysis for individual PTI genes as continuous variables across nine cancer types in which the PTI signature is prognostic.     ',\n",
       " '58Pro-tumourigenic inflammation correlates with oncogenic transformation and advancing disease stage  ',\n",
       " 'We next questioned whether qualitative changes in the COX-2 signalling network in tumour tissue are associated with quantitative increases in the expression of the PTI signature with progression from heathy to malignant diseased states.',\n",
       " 'To address this question we consulted a dataset comprising gene expression data from nine developmental stages of lung squamous cell carcinoma (LUSC) (Mascaux et al., 2019).',\n",
       " 'Unlike other studies of healthy versus cancerous tissue, this unique and powerful dataset contains samples from several histological stages of the disease enabling a more detailed analysis of changes in tissue gene expression over time prior to the onset of full carcinogenesis.',\n",
       " 'Notably, PTI levels were lowest in normal lung tissue and progressively increased (Figure 1.9A).',\n",
       " 'Elevated expression of the PTI signature was first notable in moderately dysplastic lesions and peaked in carcinoma samples (Figure 1.9A).',\n",
       " 'These findings suggest that highly proliferative, pre-cancerous lesions begin to co-opt an inflammatory transcriptional network prior to full malignancy.',\n",
       " 'Cancers can be sub-divided into stages I-IV using the TNM cancer staging criteria that take into account the size of the primary tumour, lymph node spread and the presence of metastases in other organs.',\n",
       " 'We investigated whether advancing disease stage was associated with PTI levels in the primary tumour considering the reported pro-metastatic role of tumour-elicited inflammation (Greten and Grivennikov, 2019).  ',\n",
       " 'Fitting a linear regression model on TCGA pan-cancer data with cancer type as a covariate, we found a modest but statistically significant association of PTI expression with tumour stage (P= 0.005).',\n",
       " 'In select cancer types there was a marked increase in PTI levels alongside disease progression (Figure 1.9B).',\n",
       " 'In most malignancies, PTI escalated in pre-metastatic tumour states (I-III) while in LUSC, PTI levels were comparable between stages I-III, and elevated in stage IV tumours with distant metastases (Figure 1.9B).',\n",
       " 'To conclude, PTI correlates with oncogenic transformation, neoplasia and progression to more advanced and metastatic disease stages.',\n",
       " 'Collectively this data supports the notion that inflammation is integral towards tumour progression, and establishes the PTI signature as a broad indicator of poor prognosis.         ',\n",
       " '59Figure 1.9: Pro-tumourigenic inflammation correlates with oncogenic transformation and advancing disease.',\n",
       " '(A) Mean PTI signature scores (±SEM) for samples from different stages of LUSC.',\n",
       " '(B) Correlation of disease stage with PTI score (mean ± SEM) in LUSC (n= 498), LUAD (n= 509), PAAD (n= 176), ccRCC (n= 531), and TGCT (n= 139)',\n",
       " 'TCGA datasets.    ',\n",
       " 'DISCUSSION  Accurately predicting cancer patient prognosis is a major challenge that, despite major advances in techniques for molecular biology, still relies predominantly on clinical observations, imaging techniques and systemic factors.',\n",
       " 'The role of CD8+ T cells in coordinating antigen-specific and bystander anti-tumour responses ensures that their presence in tumour compartments is, more often than not, a sign of better prognosis (Galon et al., 2006; Fridman et al., 2012).',\n",
       " 'Our data indicated that tumours with comparable levels of CD8+ T cell infiltration, IFNg signalling or TILs can have markedly different outcomes depending on the status of COX-2/PGE2-centred pro-tumourigenic inflammation in the tumour.  ',\n",
       " 'To study pro-tumourigenic inflammation we focussed our bioinformatics pipeline on signalling factors that associate with COX-2 mRNA expression.',\n",
       " 'Selection of COX-2 as a ‘hub’ gene was based on our previous findings in mouse and human (Zelenay et al., 2015; Bonavita et al., 2020).',\n",
       " 'We hypothesise that unbiased, data-driven rather than knowledge-driven, methods of ‘hub’ gene detection comparing healthy, early neoplastic and cancerous tissues would identify COX-2 as a dominant  ',\n",
       " '60regulator of the early tumour inflammatory phenotype.',\n",
       " 'To what extent the COX-2 co-expression network is overlapped with other immune evasive pathways is unclear.',\n",
       " 'Deciphering the nodal regulators of pro-tumourigenic inflammation in malignant cells themselves and separately in the stromal compartment using spatial transcriptomics  (Rodriques et al., 2019;',\n",
       " 'Merritt et al., 2020) will potentially uncover novel, and targetable, tumour-intrinsic and -extrinsic regulators of inflammation that are otherwise overlooked using bulk RNAseq approaches.  ',\n",
       " 'Patients with worse OS showed higher intratumoural PTI levels.',\n",
       " 'This was the case in a diverse set of cancers including those in the brain, kidney and pancreas each with differing sets of driver gene mutations and SCNAs.',\n",
       " 'This suggests that aggressive cancers from diverse backgrounds conscript common inflammatory processes with overlapping molecular properties to drive progression.',\n",
       " 'The correlation of PTI with disease stage was less pronounced in some malignancies.',\n",
       " 'In CESC and LIHC, for example, there was no correlation of PTI and stage, but both parameters were significantly prognostic in a multivariate Cox regression model (not shown).',\n",
       " 'Uneven and low numbers of patients across disease stage groups made it difficult to perform stage-level analyses of PTI as a prognostic factor.',\n",
       " 'Future efforts should seek to address the prognostic power of PTI in more homogeneous populations particularly in early disease settings.',\n",
       " 'We speculate that different cancer types engage pro-tumourigenic inflammatory signalling at unique time points during disease progression.  ',\n",
       " 'The observed increase in PTI levels in pre-malignant stages of LUSC suggests a role for PTI in early tumour formation yet whether PTI is co-opted to drive immune evasion or other cancer hallmarks remains unclear and is challenging to address.',\n",
       " 'Unbiased analysis of normal, chronic inflamed, and malignant tissues from hepatitis-induced liver cancers showed that angiogenesis might underpin the transition from inflammation to cancer (Chen et al., 2019).',\n",
       " 'This analysis of chronic inflammation-associated cancers suggested that macrophages and neutrophils infiltrate pre-malignant lesions during dysplasia at which point activated NK cells and CD8+ T cells are absent (Chen et al., 2019).',\n",
       " 'Thus, neoplastic cells must evolve the ability to manipulate innate inflammatory cells early on, but the exact timing at which tumour immunity is successfully subverted remains unclear.',\n",
       " 'Considering our murine findings in which it was essential to suppress NK cell activity at a timepoint prior to the development of an extensive tumour vasculature (Bonavita et al., 2020) we speculate that immune evasion may begin very early after oncogenic transformation, preceding the requirement for neo-vascularisation, and continuing through tumour progression.',\n",
       " 'Our multivariate Cox regression analysis studying the interaction between PTI and IFNg signalling indicates that when T cell activation is high, such as in tumours more likely to benefit from ICB, a concomitant increase in PTI negates the survival benefit associated with the presence of cytotoxic T cells.',\n",
       " 'We therefore conclude that antagonism of IFNg signalling or T cell inflammation plays a major role in the correlation of PTI with poor survival outcomes.',\n",
       " 'In a later chapter we will address whether this is also the case in patients treated with ICB.     ',\n",
       " '61SUPPLEMENTARY TABLES AND FIGURES  Pan-cancer PTI, TIS and interaction term survival model with disease stage and cancer type strata.',\n",
       " 'Patients = 6560 Events = 1983  Variable Comparison HR Lower 95% ConfInt Upper 95% ConfInt P value PTI Continuous 1.27 1.15 1.39 9E-07 TIS Continuous 0.81 0.76 0.87 1E-09 PTI:',\n",
       " 'TIS Continuous 1.1 1 1.2 0.04  Supplementary Table 1.1: Pan-cancer survival interaction between PTI and T cell inflammation.  ',\n",
       " 'Pan-cancer PTI, CD8 and interaction term survival model with disease stage and cancer type strata.',\n",
       " 'Patients = 6560 Events = 1983  Variable Comparison HR Lower 95% ConfInt Upper 95% ConfInt P value PTI Continuous 1.24 1.13 1.36 4E-06 CD8 Continuous 0.86 0.81 0.91 2E-07 PTI:',\n",
       " 'CD8 Continuous 1.09 1 1.18 0.04  Supplementary Table 1.2: Pan-cancer survival interaction between PTI and CD8 signatures.  ',\n",
       " 'Pan-cancer ‘Wound healing’ immune subtype patients Patients = 2090 Events = 712  Variable Comparison HR Lower 95% ConfInt Upper 95% ConfInt P value PTI Continuous 1.45 1.28 1.64 3E-09 IFNg Continuous 0.87 0.75 1.01 0.08 PTI:',\n",
       " 'IFNg Continuous 1.21 1 1.46 0.047  Supplementary Table 1.3: Survival interaction of PTI and IFNg signalling in wound healing tumours.  ',\n",
       " 'Pan-cancer ‘Lymphocyte-depleted’ immune subtype patients Patients = 1184 Events = 452  Variable Comparison HR Lower 95% ConfInt Upper 95% ConfInt P value PTI Continuous 2.07 1.67 2.56 2E-11 IFNg Continuous 0.91 0.75 1.12 0.38 PTI:',\n",
       " 'IFNg Continuous 0.77 0.57 1.04 0.09  Supplementary Table 1.4: Survival interaction of PTI and IFNg signalling in lymphocyte depleted tumours.',\n",
       " '62      Supplementary Figure 1.1: PTI combines with immune biomarkers to stratify patients with divergent outcomes.',\n",
       " '(A) KM plots of ‘Wound healing’, ‘Inflammatory’ and ‘Lymphocyte-depleted’ subtype patients stratified concomitantly by PTI and IFNg signatures.',\n",
       " '(B) OS of ‘IFNg -dominant’ patients stratified by PTI and clockwise by TIS and CD8 signatures, TIL fraction and CD274 expression.  ',\n",
       " 'For all survival analyses individual FDR values from Wald tests are shown (*FDR<0.05, **FDR<0.01, ***FDR<0.001).              ',\n",
       " '63      Chapter 2 MAPK pathway alterations and NFkB activation drive pro-tumourigenic inflammation     ',\n",
       " '64ABSTRACT  Cancer-promoting inflammation represents a prosperous environment for malignant cells to proliferate, invade, metastasise and evade innate and adaptive immune destruction.',\n",
       " 'Yet, it is unclear why not all tumours display equal flavours of inflammation.',\n",
       " 'Harnessing genomic, transcriptomic, proteomic and metabolomic data from cancer cell lines and patients alike, we explore the drivers of COX-2-centred pro-tumourigenic inflammation.',\n",
       " 'We find that cancer cell lines frequently co-express inflammatory ligands suggesting tumour cells as a main source of PTI factors, and common upstream regulation.',\n",
       " 'MEK inhibitors profoundly diminished the production of PTI mediators such as IL-6, IL-8 and IL-1a by cancer cells, an effect that was observed prior to their arrest in proliferation.',\n",
       " 'Further in silico analysis showed that PTI levels are associated with MAPK pathway alterations, most notably KRAS mutations as well as NFkB activation.',\n",
       " 'Collectively, our findings suggest that oncogenic events commonly associated with increased MAPK signalling converge on the NFkB pathway to drive the pro-tumourigenic inflammatory potential of cancer cells and highlight MEK inhibitors as candidate modulators of tumour-inflammation to raise cancer immunogenicity.                             ',\n",
       " '65INTRODUCTION  Cancer cells secrete a plethora of different soluble factors including lipids, proteins and peptides among other molecules.',\n",
       " 'A large proportion of these molecules signal directly via cell surface receptors or indirectly through their involvement in enzymatic processes and biological pathways.',\n",
       " 'Autocrine and/or paracrine signalling between cancer cells directly aids survival, proliferation, EMT and can promote resistance to therapy (Madden et al., 2020).',\n",
       " 'Such signalling establishes positive feedback loops to enhance production of other secretome factors (Kulbe et al., 2007).',\n",
       " 'Thus, secretion of inflammatory mediators by cancer cells may depend on continual ligation by other factors in addition to cell autonomous oncogenic signalling.',\n",
       " 'Paracrine signalling between cancer cells and non-cancer cells also promotes tumour progression.',\n",
       " 'Fibroblasts residing in the intestinal mesenchyme produce PGE2 that elicits expansion of tumour cell progenitors via EP4 receptor signalling thereby facilitating cancer initiation (Roulis et al., 2020).',\n",
       " 'In the reverse direction, lung cancer cells secrete PGE2 that acts on local stromal cells to induce prolactin secretion and subsequent tumour formation (Zheng et al., 2019).',\n",
       " 'Cancer cells also produce endocrine factors capable of long-range signalling and promotion of systemic inflammation.',\n",
       " 'For example, IL-1b and G-CSF collaborate to elevate bone marrow production of pro-metastatic neutrophils (Coffelt et al., 2015).',\n",
       " 'Our previous data showed how murine cancer models routinely express COX-2 and produce PGE2 that is capable of diverse signalling within the TME (Zelenay et al., 2015; Bonavita et al., 2020; Pelly et al., 2021), yet whether PGE2 acts in concert with other inflammatory mediators is unclear.',\n",
       " 'Although several publications have reported links between oncogenic signalling and production of immunomodulatory factors, the majority of these focus on one or two mediators rather than transcriptional programs (Shchors et al., 2006; Kumar et al., 2012; Spranger, Bao and Gajewski, 2015; Koyama et al., 2016; Peng et al., 2016; McGrail et al., 2018; Wellenstein et al., 2019).',\n",
       " 'In addition, it remains unclear how tumour cells upregulate the COX-2 pathway.',\n",
       " 'Reports have suggested both tumour-intrinsic (Markosyan et al., 2019) and -extrinsic (Che et al., 2017) mechanisms of pathway activation yet there are no systematic, pan-cancer analyses of the genomic correlates.',\n",
       " 'Having identified the PTI signature as being powerfully prognostic and antagonistic of T cell inflammation in human cancer we sought to investigate cancer cell-intrinsic expression of the genes.',\n",
       " 'By performing a deep interrogation of publicly available multi-omics data from cancer cell lines, we aimed to establish an explanatory model by which cancer cells are able to produce inflammatory mediators, and in turn uncover preceding evolutionary events, putative drivers and potentially targetable nodes.      ',\n",
       " '66RESULTS  Cancer cell lines expressing pro-tumourigenic inflammatory mediators are sensitive to MEK inhibition Considering the broad prognostic utility of the PTI signature and the highly coordinated network interaction of PTI factors in tumour tissue, we hypothesised that cancer cells themselves would produce many of these molecules.',\n",
       " 'Previously we showed how numerous murine cancer cell lines express COX-2 and secrete PGE2, however it is unclear the extent to which cancer cells co-express other inflammatory modulators (Zelenay et al., 2015).',\n",
       " 'To address this hypothesis, we analysed proteomic, genomic, metabolomic, transcriptomic and gene essentiality data from the cancer cell line encyclopaedia (CCLE) (McFarland et al., 2018; Ghandi et al., 2019; Nusinow et al., 2020).',\n",
       " 'This deep and integrated analysis harnessed data from more than 800 human solid tumour cell lines.',\n",
       " 'Altogether, there were cell lines with varying in vitro morphology and proliferation rate including 12% non-small cell lung cancer (NSCLC), 7% colorectal and 6% breast cancer samples derived from both primary and metastatic (29%) tumours.',\n",
       " 'PTI genes were expressed heterogeneously amongst the cell lines, and co-correlation analysis showed how these ligands were highly correlated with one another (Figure 2.1A).',\n",
       " 'Indeed, the mean pairwise Spearman correlation coefficient of PTI genes was 0.36 (0.11-0.6) that was substantially higher than the maximum, mean correlation of 1000 randomly selected gene sets (Spearman= 0.15).',\n",
       " 'Therefore, we concluded that cancer cells themselves express PTI genes in concert as part of a transcriptional program.  ',\n",
       " 'We subsequently calculated cancer cell-intrinsic PTI scores in order to correlate expression of this transcriptional program with cell line characteristics.',\n",
       " 'Cancer cell-intrinsic PTI score was not powerfully correlated with patient age (Spearman = 0.17, P= 5E-05), mutation rate (Spearman= -0.07, P= 0.07), or cell line doubling time (Spearman= -0.1, P= 0.04).',\n",
       " 'Males had marginally higher PTI scores (P= 0.05), and MSI cell lines had significantly lower PTI scores (P= 0.0002) (Supplementary Figure 2.1).',\n",
       " 'PTI scores were also heterogeneous between tissue types with no obvious enrichment for cancers in which there is a clear link with chronic inflammation (LIHC, stomach (STAD) and colorectal cancer (COADREAD))',\n",
       " 'amongst those with the highest PTI levels (Supplementary Figure 2.1).',\n",
       " 'Other than trends between tissue types, clinical characteristics did not powerfully predict expression of inflammatory mediators by cancer cells.      ',\n",
       " '67 Figure 2.1: Cancer cell lines co-express pro-tumourigenic inflammatory ligands.',\n",
       " '(A) Pairwise Spearman correlation coefficients of PTI genes in non-haematological, non-fibroblast cancer cell lines from the CCLE.',\n",
       " 'Non-significant (P> 0.05) relationships are marked by ’X’.',\n",
       " '(B) Heatmap illustrating the stratification of PTI high and low cancer cell lines from the Garnett et al (2012) drug screen dataset, which comprises microarray gene expression data for more than 600 cell lines.',\n",
       " 'These cell lines partially overlap those in the CCLE dataset.  ',\n",
       " 'To search for putative ways of targeting cancer cells that produce PTI factors we first analysed data from a pan-cancer drug screen in which over 600 human cancer cell lines, with matched transcriptomes, were treated with a panel of chemotherapies and targeted agents (Garnett et al., 2012).',\n",
       " 'We hypothesised that cell lines secreting PTI factors would be selectively sensitive to compounds that target their upstream pathways.',\n",
       " 'To investigate this, we first grouped cell lines by PTI expression, defining PTIhigh cell lines as those in the top quantile (Figure 2.1B).',\n",
       " 'All other cell lines were defined as PTIlow.',\n",
       " 'This allowed us to screen for associations between PTI status and drug sensitivity whilst adjusting for cancer type.',\n",
       " 'PTIhigh lines were selectively sensitive to four different MEK inhibitors (refametinib, CI1040, PD0325901, AZD6244) as well as inhibitors of EGFR (erlotinib), and cyclin-dependent kinases (CDK) (CGP60474) (Figure 2.2A).',\n",
       " 'On the contrary, PTIlow cell lines were sensitive to inhibitors of PARP1/2 and receptor tyrosine kinase (RTK) inhibitors that target VEGFR, PDGFR and FGFR (Figure 2.2A).',\n",
       " 'Therefore, PTIhigh cell lines are not susceptible to blanket inhibition of RTKs the majority of which are thought to signal via ERK.',\n",
       " 'In fact, more targeted inhibition of the MAPK cascade with MEK1/2 inhibitors led to potent inhibition of PTIhigh but not PTIlow cell lines.',\n",
       " 'As an example, for cell lines treated with MEK1/2 inhibitor refametinib, there was a significant anti-correlation between PTI score and refametinib half-maximal inhibitory concentration (IC50) (Figure 2.2B, top panel).',\n",
       " 'Refametinib-sensitive PTIhigh lines in the top left quadrant (Figure 2.2B, top panel) were derived from several cancer types including 27% NSCLC, 18% oesophageal, 8% pancreas and 8% bladder.',\n",
       " 'Using chi-squared (c2) tests, we checked whether there was enrichment of cell lines derived from a specific cancer type amongst the refametinib-sensitive, PTIhigh group, and found only that these lines contained significantly more NSCLCs compared to all others (21% v 7%, FDR= 0.006).',\n",
       " 'To test whether the increased MEK inhibitor sensitivity of PTIhigh cell lines was confounded by this cancer type over-representation, we repeated the analysis with only NSCLC cell lines which revealed an even more  ',\n",
       " '68pronounced anti-correlation of refametinib IC50 and PTI score (Figure 2.2B, bottom panel).',\n",
       " 'Indeed, examining the dose response curves for refametinib in ten PTI high and low NSCLC cell lines there was a profound shift in which PTIlow cell lines required far higher concentrations of drug to achieve equal levels of growth inhibition (Figure 2.2C).',\n",
       " 'Interestingly, refametinib-sensitive PTIhigh cell lines were not enriched in melanomas (MEL).',\n",
       " 'On the contrary, amongst all refametinib-sensitive cell lines, MELs were enriched in PTIlow (20/72, 28%) rather than PTIhigh cell lines (2/49, 4%) (c2, P= 0.002).',\n",
       " 'Thus, cell lines expressing PTI factors that were efficiently killed by MEK inhibitors are from cancer types in which MEK inhibitors are not yet approved for clinical use.       ',\n",
       " 'Figure 2.2: Cancer cell lines expressing high levels of pro-tumourigenic inflammatory mediators are sensitive to MEK inhibition.',\n",
       " '(A) PTI status coefficients from linear mixed effects models in which PTI status (low/ high) was used to predict cell line drug sensitivity adjusted for cancer type.',\n",
       " 'Drug names (black) and targets (blue/red) are labelled on the ‘y’ axis.',\n",
       " 'All bars shown are from statistically significant associations (P< 0.05).',\n",
       " '(B) Correlation of cell line sensitivity to refametinib by the natural log of the drug IC50 value, and of the cancer-intrinsic PTI transcriptional program for all cell lines (top), and for NSCLC lines (bottom).',\n",
       " '(C) Dose response curves of ten PTIhigh and ten PTIlow refametinib–treated NSCLC cell lines.',\n",
       " 'Shapes represent the different cell lines.    ',\n",
       " 'Deleted: 2',\n",
       " '69To determine whether pharmacological inhibition of MEK directly supresses the production of PTI factors we obtained two human lung cancer cell lines (NCIH1339, NCIH3255) with high PTI levels according to the CCLE RNAseq dataset.',\n",
       " 'Trametinib, an approved and selective MEK1/2 inhibitor, suppressed proliferation of these cells 24 hours post-treatment (Figure 2.3A-B).',\n",
       " 'In line their transcriptome, analysis of their supernatants by multiplex cytokine/chemokine assays showed that both cell lines produced PTI factors IL-1α, CXCL1, IL-6, and IL-8 (Figure 2.3C).',\n",
       " 'Notably, as early as 6 hours and for as long as 24 hours post-treatment prior to the observed effects on cell proliferation, trametinib inhibited the production of all these inflammatory mediators by both the KRASG12R (NCIH1339) and KRASWT EGFR-amplified cell lines (NCIH3255).',\n",
       " 'These results were mimicked by a second MEK1/2 inhibitor selumetinib (Supplementary Figure 2.1).',\n",
       " 'Importantly, the inflammatory mediator osteopontin, which is not part of the PTI program, was also produced by NCIH1339 cells but its levels were not altered following MEK inhibition (Supplementary Figure 2.1) indicating that MEK inhibition does not indiscriminately inhibit cytokine production.',\n",
       " 'Collectively this data suggests that inhibition of MEK kinase activity can rapidly block the production of PTI factors by cancer cells.   ',\n",
       " '70 Figure 2.3: MEK inhibitors attenuate secretion of pro-tumourigenic inflammatory mediators prior to anti-proliferative effects.  ',\n",
       " '(A) Change from baseline confluence percentage over time for cell lines NCIH1339 (top) and NCIH3255 (bottom) treated with two concentrations of trametinib in 96 well plates.',\n",
       " 'Lines indicate timepoints of supernatant analysis.',\n",
       " '(B) Representative images from confluence readouts after 8, 24 and 72 hours of 10μM trametinib treatment.',\n",
       " 'Confluence percentage of the entire well is shown in red.',\n",
       " '(C) Concentration of secreted proteins detected by Luminex 6-, 8- and 24-hours following treatment of NCIH1339 (top panel) and NCIH3255 (bottom row) cells with trametinib at 1μM and 10μM.  ',\n",
       " '71MAPK and NFκB activation associates with PTI expression in cancer cell lines and tumours  Considering the differential sensitivity of PTIhigh cell lines to MEK inhibitors, we hypothesised that oncogenic alterations, either mutations and/or SCNAs, associated with increased MAPK activity, would correlate with PTI expression.',\n",
       " 'Consistent with this hypothesis, KRAS mutated cell lines had higher PTI expression independently of cancer type and mutation rate (Figure 2.4A).',\n",
       " 'KRAS mutations occur at specific codon hotspots such as positions 12, 13 and 61 (Figure 2.4B).',\n",
       " 'There was a consistent upregulation of PTI in cell lines grouped by either of the three most frequent hotspot mutations (Figure 2.5B).',\n",
       " 'Moreover, cell lines with KRASG12C mutations that are most prevalent in NSCLC and KRASG12D mutations, most frequent in pancreatic and colorectal tumours, were individually associated with higher PTI expression (Supplementary Figure 2.3).',\n",
       " 'These findings suggest that the association of KRAS mutations with PTI expression is also independent of the specific amino acid alteration.',\n",
       " 'We therefore conclude that oncogenic KRAS signalling is a strong pan-cancer driver of PTI through cancer cell-intrinsic transcriptional activation.',\n",
       " 'Next, we addressed whether gene-level focal SCNAs were predictive of high PTI expression using multivariate logistic regression modelling.',\n",
       " 'For this, we defined PTIhigh CCLE cell lines as those in the top quantile, and PTIlow as those in the bottom quantile.',\n",
       " 'Given that aneuploidy score had a minor positive correlation with cancer cell-intrinsic PTI score (Spearman= 0.11, P= 0.003) we adjusted for both cancer type and aneuploidy in each model.',\n",
       " 'Amongst the genes whose copy number estimates were significantly predictive of high PTI expression were HRAS, RASA1, a RAS GTPase activating protein (GAP), RTKs PDGFRB and FLT4 that signal via MAPK signalling, RELA, the gene encoding for NFκB subunit p65 that can be trans-activated by MEK, and a subunit for the IL-6 receptor (IL6ST) (Figure 2.4D)',\n",
       " '(Chang et al., 2003).',\n",
       " 'These results suggest that cancer cells expressing high levels of inflammatory mediators carry KRAS mutations but also amplifications in the MAPK-NFκB cascades that may equally promote PTI expression.            ',\n",
       " '72  Figure 2.4: KRAS mutations, MAPK alterations and NFκB pathway amplifications associate with high cancer cell-intrinsic PTI expression.',\n",
       " '(A) Boxplot showing the association of the cancer cell-intrinsic PTI transcriptional program and KRAS mutations in pan-cancer cell lines.',\n",
       " '(B) A lollipop plot showing the KRAS amino acid sequence alongside the number of mutations at each codon shown on the y axis.',\n",
       " '(C) Comparison of cancer cell-intrinsic PTI scores between cell lines with wild-type KRAS (WT), and those with mutations at different amino acid residues.',\n",
       " '(D) Copy number estimates for five different genes separated by cancer cell-intrinsic PTI status.',\n",
       " 'A value of zero indicates diploid copies of the gene.  ',\n",
       " 'The presence of IL6ST amplifications in PTIhigh cell lines known to secrete IL-6 suggests that autocrine signalling by PTI factors may be important for the growth and/or survival of these cell lines.',\n",
       " 'To further investigate the role of cancer cell-intrinsic expression of the receptors for PTI factors, we analysed integrated data from multiple short hairpin RNA (shRNA)-based screens for gene essentiality (McFarland et al., 2018).',\n",
       " 'Using this gene dependency dataset, we questioned whether PTIhigh cell lines were selectively sensitive to loss of PTI ligand receptors allowing us to infer potential autocrine signalling nodes within the PTI transcriptional program (Figure 2.5A).',\n",
       " 'Knockdown of PTGER4 and EGFR led to cell death selectively in PTIhigh cancer lines compared to PTIlow (Figure 2.5A, B).',\n",
       " 'Consistent with our earlier hypothesis, IL6ST was also an essential gene in PTIhigh cancer lines, albeit at a lower level of significance (Figure 2.5A).',\n",
       " 'This implies autocrine or paracrine signalling by PTI factors in cancer cell survival.',\n",
       " 'Furthermore, in PTIhigh cells transcription factor TEAD1 that promotes signalling via YAP/ TAZ, RAC1, a GTPase similar to KRAS, and TLR5, an innate immune receptor, were among the top  ',\n",
       " '73ranked, most essential genes (Figure 2.5C).',\n",
       " 'Collectively, these results highlight how cell lines overexpressing inflammatory ligands have distinct wiring to cell lines that have low PTI expression, and are thereby dependent on unique sets of genes for survival.      ',\n",
       " 'Figure 2.5:',\n",
       " 'The survival of PTI producing cancer cell lines relies on EGFR and PTGER4 expression.',\n",
       " '(A) Volcano plot showing results from logistic regression models predicting cancer cell-intrinsic PTI status (high/ low) by gene essentiality scores.',\n",
       " 'Model coefficients are on the x axis and -log10 P values on the y axis.',\n",
       " 'Negative coefficients indicate that the gene is essential for survival of the particular cell line.',\n",
       " '(B) Comparison of EGFR and PTGER4 gene essentiality scores between PTI high, low and intermediate cell lines.',\n",
       " 'Negative values indicate that loss is associated with cell death.',\n",
       " 'Wilcoxon test P values are shown.',\n",
       " '(C) Table showing the top 15 most essential genes in PTI high and low expressing cell lines.',\n",
       " 'FDR values from logistic regression models are shown.   ',\n",
       " 'Given that MAPK and NFκB pathways intersect, we hypothesised that convergence around NFκB activation was likely to be the dominant driver of PTI levels.',\n",
       " 'Consistent with this hypothesis, analysis of reverse phase protein array (RPPA) data showed that KRAS mutant cell lines with high PTI had significantly increased levels of activating p65 (RELA) serine 536 phosphorylation, in addition to higher protein levels of EGFR, ECM components fibronectin and transglutaminase, as well as annexin-1 that regulates NFκB activity (Moraes, Ampomah and Lim, 2018)',\n",
       " '(Figure 2.6A, B).',\n",
       " 'These findings at the protein level suggest that NFκB activity promotes expression of the PTI transcriptional program in cell lines with KRAS mutations.  ',\n",
       " '74The strong positive correlation between NFκB activity and cancer cell-intrinsic PTI scores led us to hypothesise that p65/RELA must directly regulate transcription of PTI genes by binding to their cis-regulatory, promoter regions.',\n",
       " 'To address this hypothesis, we searched for common regulatory sequences within 500 base pairs upstream of the transcription start site (TSS) of PTI genes.',\n",
       " 'To increase the power of our analysis we supplemented the PTI gene list with an additional set of 485 genes that were highly and significantly upregulated in PTIhigh versus PTIlow cell lines (FC> 5, FDR< 0.05).',\n",
       " 'In agreement with our previous data, there were three NFκB transcription factors, including highly conserved motifs that are bound both by canonical (NFKB1) and non-canonical forms (NFKB2) of NFκB, amongst the top ten motifs significantly enriched in the promoter regions of PTI genes (Figure 2.6C-D).',\n",
       " 'Thus, motif enrichment analysis supports the stimulatory role of NFκB signalling in cancer cell-intrinsic PTI ligand expression.    ',\n",
       " 'Figure 2.6: NFκB binds the promoter regions of PTI genes and promotes their expression.',\n",
       " '(A) Volcano plot showing differential expression of proteins and phosphoproteins measured by RPPA comparing PTIhigh and PTIlow KRAS mutant cell lines.',\n",
       " '(B) Levels of p65/RELA pS536 in RPPA data from KRAS mutated cancer cell lines stratified according to cancer-intrinsic PTI status.',\n",
       " '(C) Table showing the top ten significantly enriched motifs in the extended, cancer cell-intrinsic PTI gene list alongside predicted transcription factor binders.',\n",
       " '(D) Sequence logos from multiple sequence alignments of NFKB1 and NFKB2 binding motifs from chromatin immunoprecipitation sequencing (ChIPseq) experiments showing consensus nucleotide sequences for the associated motifs.      ',\n",
       " '75To test whether intratumoural PTI levels correlate with genomic alterations that lead to MAPK and NFκB pathway overactivation we analysed mutation calls from the pan-cancer TCGA dataset.',\n",
       " 'On a pan-cancer basis, mutations in MAPK pathway components, MAP2K1, that encodes MEK1, NF1 and BRAF were significant predictors of PTI expression independent of mutation burden and cancer type (FDR< 0.25).',\n",
       " 'TP53 mutations were also significantly positively correlated with PTI levels (Figure 2.7A).',\n",
       " 'In contrast, oncogene (NFE2L2) and tumour suppressor gene mutations (AXIN2, KEAP1) in NRF2 and WNT pathways were associated with generally lower levels of PTI (Figure 2.7A).',\n",
       " 'We therefore deduce that there is specificity in the link between MAPK cascade mutations, pathway activation and tumour-promoting inflammation.',\n",
       " 'The same analysis in individual cancer types yielded 33 significant mutational links to high PTI expression (FDR< 0.25).',\n",
       " 'Of these mutations, a large fraction were MAPK pathway genes including KRAS in LUAD, ccRCC and PAAD, NF1 in lung squamous cell carcinoma (LUSC) and GBM, and BRAF in THCA (Figure 2.7B).',\n",
       " 'As such, MAPK activation is a hallmark of PTIhigh tumours in multiple tissue types.',\n",
       " 'Moreover, we found that EGFR and RELA amplifications in solid tumours were positively correlated with intratumoural PTI signature scores (Figure 2.7C).',\n",
       " 'As a result, we conclude that MAPK/ NFκB pathway mutations and SCNAs promote both cancer cell-intrinsic and tumour microenvironmental expression of PTI factors.                        ',\n",
       " '76  Figure 2.7: MAPK pathway mutations and RELA amplifications associate with high levels of pro-tumourigenic inflammation.',\n",
       " '(A) Odds ratios and 95% confidence intervals from multivariate linear regression models using mutation status to predict PTI scores in the pan-cancer TCGA dataset adjusted for cancer type and TMB.',\n",
       " '(B)',\n",
       " 'The proportion of significant mutation-to-PTI links that are in different oncogenic signalling pathways.',\n",
       " 'Breakdown of significant MAPK pathway mutation-to-PTI links in individual cancers.',\n",
       " '(C) Comparison of intratumoural PTI signature scores between pan-cancer TCGA tumours with diploid copies of EGFR and RELA (WT) versus those with chromosomal amplifications.',\n",
       " 'Wilcoxon test P values are shown.   ',\n",
       " 'Pro-tumourigenic inflammation correlates with additional immune evasive features of cancer cells  We next sought evidence of potential additional immune evasive features of cancer cells with high PTI expression that may contribute towards the profound antagonism of the PTI signature towards T cell inflammation in human tumours.',\n",
       " 'Firstly, we analysed the results of DGE analysis between PTI high and low expressing cell lines.',\n",
       " 'There was a consistent trend for PTIhigh cancer cells to have higher transcript abundance for MHC class I (HLA-A/B/C) and II (HLA-DRB5/DRA) molecules as well as MICB, a stress ligand that binds the activation receptor NKG2D and can stimulate T cell and NK cell cytotoxicity (Figure 2.8A).',\n",
       " 'These findings imply that PTIhigh cancer cells have greater propensity for antigen  ',\n",
       " '77presentation than PTIlow counterparts.',\n",
       " 'Upregulation of CD274, the gene encoding PD-L1, to the same extent as individual PTI elements indicates the high degree of overlap between the PTI signature and PD-L1 expression by cancer cells (Figure 2.8A).',\n",
       " 'In addition to PD-L1, PTIhigh cell lines had elevated expression of other genes similarly implicated in immunosuppressive signalling by cancer cells.',\n",
       " 'These include IDO1 that encodes indoleamine-pyrrole 2,3-dioxygenase (IDO), an enzyme that converts tryptophan to kynurenine, and NT5E that encodes 5’nucleotidase (NT5E), an enzyme that converts adenosine monophosphate (AMP) to adenosine.',\n",
       " 'Both of these enzymes are thought to have immunosuppressive functions in the TME (Munn and Mellor, 2016; Vijayan et al., 2017).',\n",
       " 'Analysing cancer cell line metabolomics data, we found that cellular levels of IDO product kynurenine were associated with increased PTI expression as well as NT5E product adenosine, whilst AMP was elevated in PTIlow cells (Figure 2.8B).',\n",
       " 'Thus, metabolomics data supports the notion that cancer cells capable of secreting abundant PTI mediators, also favour metabolic pathways that generate putative immunosuppressive metabolites.',\n",
       " 'Finally, using cancer cell line mass spectrometry proteomics data we questioned whether the association of the PTI transcriptional program with other immune evasive features of cancer cells was maintained at the protein level.',\n",
       " 'Whilst IDO1 levels were not evaluable, there were strong positive correlations of the PTI signature with PD-L1 and NT5E levels consistent with observations at the mRNA level (Figure 2.8C).',\n",
       " 'These findings therefore extend the potential of cancer cells that overexpress pro-tumourigenic inflammatory ligands to sculpt a TME that restrains tumour immunity.                             ',\n",
       " '78 Figure 2.8: Cancer cells expressing pro-tumourigenic inflammatory mediators have high levels of PD-L1 and immunosuppressive metabolites.',\n",
       " '(A) Log2FC values for selected genes from the comparison of PTI high and low cancer cell line transcriptomes.',\n",
       " 'Points coloured grey represent those with a non-significant FDR (>0.05).',\n",
       " '(B) Results from logistic regression models using metabolite levels to predict PTI high versus low cell lines whilst adjusting for cancer type.',\n",
       " 'Model coefficients are represented on the x axis and -log10 normalised FDR values on the y axis.',\n",
       " '(C) Correlation of cancer cell-intrinsic PTI scores with protein levels of PD-L1 and NT5E detected by mass spectrometry-based proteomics.',\n",
       " 'Spearman Rho and P values are shown.  ',\n",
       " 'DISCUSSION  Cancers with unique tissue origin, cell-of-origin, morphology, and genetic alterations have distinct transcriptomics that are recapitulated by human cancer cell lines (Ghandi et al., 2019; Nusinow et al., 2020).',\n",
       " 'Our results highlighted a distinct set of these cell lines with elevated expression of inflammatory mediators such as PTGS2 and CXCL8 that are known to promote tumourigenesis and immune evasion (Zelenay et al., 2015; Teijeira et al., 2021).',\n",
       " 'The concomitant expression of these inflammatory mediators by cancer cells suggests that common upstream genetic events drive their transcription.',\n",
       " 'We set out to dissect the underlying basis for the differential secretion of PTI factors by cancer cell lines.',\n",
       " 'Cancer cells expressing the PTI transcriptional program have genetic alterations in the MAPK and NFκB pathways.',\n",
       " 'Previous studies showed correlations between these pathways and COX-2 activity but not the entire PTI transcriptional program (Hennequart et al., 2017; Markosyan et al., 2019; Feng et al., 2020).',\n",
       " 'KRAS mutations that strongly correlated PTI expression promote hostile inflammatory TMEs (Todoric and Karin, 2019).',\n",
       " 'In recent years, pharmacological inhibitors of KRASG12C have been  ',\n",
       " '79discovered.',\n",
       " 'These drugs were shown to favourably modulate the inflammatory TME in cancer models (Canon et al., 2019; Fedele et al., 2021), and were well tolerated in humans providing clinical benefit to refractory KRASG12C lung cancer patients (Skoulidis et al., 2021).',\n",
       " 'However, acquired resistance to KRASG12C inhibitors via additional KRAS mutations, as well as EGFR, PTEN, BRAF, MAP2K1 and NF1 alterations is an emerging problem (Awad et al., 2021).',\n",
       " 'Many of these mutations were also linked with PTI levels either in cancer cell lines or in tumours, suggesting that PTI may have a role in acquired resistance to targeted KRASG12C inhibitors.',\n",
       " 'Several lines of existing evidence establish a fundamental requirement for NFκB signalling in tumours with TP53 loss and KRAS mutations (Barbie et al., 2009; Meylan et al., 2009; Bassères et al., 2010; Ling et al., 2012).',\n",
       " 'Multiple oncogenic pathways converge upon NFκB signalling thereby illustrating its importance in tumour development (Taniguchi and Karin, 2018; Verzella et al., 2020).',\n",
       " 'In KRAS mutant pancreatic cancer, PTI factor IL-1a engages in an autocrine forward feedback loop to activate the NFκB pathway (Ling et al., 2012).',\n",
       " 'In KRAS mutant lung cancer lines, TP53 restoration in TP53-deficient cells, led to NFκB inhibition suggesting that TP53 status controls the NFκB pathway (Meylan et al., 2009).',\n",
       " 'Large-scale genomic studies have also highlighted a significant  enrichment for NFκB pathway focal SCNAs in human tumours (Beroukhim et al., 2010).',\n",
       " 'For example, 5% of uterine carcinosarcomas, 4% of head and neck cancers and 3% of bladder cancers in the TCGA have RELA amplifications.',\n",
       " 'Tumours may therefore adopt different routes to enhance NFκB signalling whether through autocrine signalling loops, mutations or SCNAs.',\n",
       " 'Targeting either canonical and/or non-canonical NFκB signalling may attenuate cancer cell-intrinsic PTI expression, however inhibition of such a ubiquitously important pathway is fraught with risks, not least due to potential side effects but also as a result of cell type-specific functions of the pathway (Cuzick et al., 2009; Gerondakis et al., 2014; Grinberg-Bleyer et al., 2017).',\n",
       " 'Whether there is a sufficient therapeutic window to inhibit NFκB in patients with SCNAs in the pathway remains to be seen.',\n",
       " 'In addition to cancer cell line genetic features, epigenetic factors likely contribute towards heterogeneity in PTI expression.',\n",
       " 'Indeed, tissue type strongly associated with PTI expression, and epigenetic regulation of transcription is the major determinant of tissue-specific cellular differentiation (Maston, Evans and Green, 2006; Ong and Corces, 2011).',\n",
       " 'Although NFκB consensus binding motifs were highly enriched directly upstream of the TSSs of genes upregulated in PTIhigh cell lines, transcriptional regulation can occur at more distant sites including enhancers located in different chromosomes (Ong and Corces, 2011).',\n",
       " 'Recently, pan-cancer analysis identified an enhancer 140 kilobases upstream of the PD-L1 gene that tightly regulates its expression (Chen et al., 2018).',\n",
       " 'Interestingly, ChIPseq data for NFκB p65 showed a strong peak for the same enhancer region, and for its binding motif in the PD-L1 promoter region (Chen et al., 2018).  ',\n",
       " 'Therefore, NFκB facilitates an enhancer-promoter interaction that regulates PD-L1 expression, which was highly correlated with cancer cell-intrinsic PTI gene transcription.',\n",
       " 'Utilising a published method (Jerby-Arnon et al., 2018) to analyse cell line drug screen data (Garnett et al., 2012), we showed that PTIhigh cell lines are sensitive to MEK inhibitors.',\n",
       " 'The dataset was, however, not supplemented with next generation inhibitors of historically challenging targets such as KRAS, or novel epigenetic modulators such as those targeting bromodomain-containing proteins',\n",
       " '80(Canon et al., 2019; Faivre et al., 2020).',\n",
       " 'In addition, identifying a core targetable node to disrupt the PTI signalling network will require experiments in vivo with the full complement of interacting cell types.',\n",
       " 'Nevertheless, our recent work indicated that the COX-2/PGE2 axis is a strong candidate (Pelly et al., 2021).',\n",
       " 'Numerous studies have reported the favourable, immunomodulatory properties of MEK inhibitors (Hu-Lieskovan et al., 2015; Ebert et al., 2016; Bommareddy et al., 2018; Baumann et al., 2020).',\n",
       " 'While future work is required to test if MEK inhibitors specifically dampen the PTI transcriptional program in vivo, our data further supports the notion that they have broader scope not just as cytostatic agents (Templeton and Musib, 2015; Hamarsheh et al., 2020).  ',\n",
       " 'Our data also highlights a link between the production of inflammatory mediators by cancer cells, and IDO1-mediated tryptophan metabolism that generates kynurenine; an immunosuppressive metabolite in the TME (Munn and Mellor, 2016).',\n",
       " ...]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3a66a246-c7ef-47a3-a110-5a78acc13007",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'text_0.txt'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the file name\n",
    "fname = paths[0].split('/')[-1]\n",
    "fname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d0423389-71c9-47b6-be7b-d81cc75a5e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tokenize the text in these lines\n",
    "sample = tokenizer(lines, max_length=512, padding='max_length', truncation=True, return_tensors='pt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e0a9a9ab-016e-4cc7-afd9-ef191d3dfb20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['input_ids', 'attention_mask'])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "03389908-8cc7-4297-9285-ede37eadcc54",
   "metadata": {},
   "outputs": [],
   "source": [
    "tokenized_text[fname] = sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "11ea7580-cfc4-44e3-a539-476defa39993",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the sample object contains some of our tensors - extract these\n",
    "## get the input IDs and append to labels\n",
    "labels.append(sample.input_ids)\n",
    "## get the attention mask - the binary\n",
    "mask.append(sample.attention_mask)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "18c25cb3-9560-466c-9b70-058b5a213d30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[   0, 1202,   37,  ...,    1,    1,    1],\n",
       "        [   0,   56,  309,  ...,    1,    1,    1],\n",
       "        [   0,  892,  568,  ...,    1,    1,    1],\n",
       "        ...,\n",
       "        [   0, 1543,  261,  ...,    1,    1,    1],\n",
       "        [   0, 3796,   93,  ...,    1,    1,    1],\n",
       "        [   0,  564, 2787,  ...,    1,    1,    1]])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample.input_ids.detach().clone()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d23f9978-a976-4a1b-9d4f-0750c8fab99e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## now apply the masked language model function on the input IDs to mask 15% of tokens\n",
    "mlm_on_input = masked_language_model(sample.input_ids.detach().clone())\n",
    "input_ids.append(mlm_on_input)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bd962def-f854-4e4a-866e-15cbcceecd4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# construct the output\n",
    "encodings = {\n",
    "    'input_ids': input_ids[0],\n",
    "    'mask': mask[0],\n",
    "    'labels': labels[0]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bb55b92a-20c8-4939-838c-e4e197b384d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1868"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(encodings['labels'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e179b67f-7ffc-4b08-a494-12cafc091c2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Dataset(torch.utils.data.Dataset):\n",
    "    def __init__(self, encodings):\n",
    "        self.encodings = encodings\n",
    "    def __len__(self):\n",
    "        return self.encodings['input_ids'].shape[0]\n",
    "    def __getitem__(self, i):\n",
    "        return {\n",
    "            'input_ids': self.encodings['input_ids'][i]\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7f7409d9-17b2-4293-ac00-04831a9ed69d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_data_loader(encodings):\n",
    "    dataset = Dataset(encodings)\n",
    "    data_loader = torch.utils.data.DataLoader(dataset, batch_size=16, shuffle=True)\n",
    "    return data_loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ba61d052-b11f-44b6-b0a6-5ca4ab6e400b",
   "metadata": {},
   "outputs": [],
   "source": [
    "loader = create_data_loader(encodings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "be411fa9-d271-4655-9498-882aa7a6e2b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['_DataLoader__initialized',\n",
       " '_DataLoader__multiprocessing_context',\n",
       " '_IterableDataset_len_called',\n",
       " '__annotations__',\n",
       " '__class__',\n",
       " '__class_getitem__',\n",
       " '__delattr__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__iter__',\n",
       " '__le__',\n",
       " '__len__',\n",
       " '__lt__',\n",
       " '__module__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__orig_bases__',\n",
       " '__parameters__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__sizeof__',\n",
       " '__slots__',\n",
       " '__str__',\n",
       " '__subclasshook__',\n",
       " '__weakref__',\n",
       " '_auto_collation',\n",
       " '_dataset_kind',\n",
       " '_get_iterator',\n",
       " '_index_sampler',\n",
       " '_is_protocol',\n",
       " '_iterator',\n",
       " 'batch_sampler',\n",
       " 'batch_size',\n",
       " 'check_worker_number_rationality',\n",
       " 'collate_fn',\n",
       " 'dataset',\n",
       " 'drop_last',\n",
       " 'generator',\n",
       " 'multiprocessing_context',\n",
       " 'num_workers',\n",
       " 'persistent_workers',\n",
       " 'pin_memory',\n",
       " 'pin_memory_device',\n",
       " 'prefetch_factor',\n",
       " 'sampler',\n",
       " 'timeout',\n",
       " 'worker_init_fn']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(loader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4f09ad64-2bad-4547-bd7b-bae4fa7c6db4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'input_ids': tensor([[   0, 1202,   37,  ...,    1,    1,    1],\n",
       "         [   0,   56,  309,  ...,    1,    1,    1],\n",
       "         [   0,  892,  568,  ...,    1,    1,    1],\n",
       "         ...,\n",
       "         [   0, 1543,  261,  ...,    1,    1,    1],\n",
       "         [   0, 3796,   93,  ...,    1,    1,    1],\n",
       "         [   0,  564, 2787,  ...,    1,    1,    1]]),\n",
       " 'mask': tensor([[1, 1, 1,  ..., 0, 0, 0],\n",
       "         [1, 1, 1,  ..., 0, 0, 0],\n",
       "         [1, 1, 1,  ..., 0, 0, 0],\n",
       "         ...,\n",
       "         [1, 1, 1,  ..., 0, 0, 0],\n",
       "         [1, 1, 1,  ..., 0, 0, 0],\n",
       "         [1, 1, 1,  ..., 0, 0, 0]]),\n",
       " 'labels': tensor([[   0, 1202,   37,  ...,    1,    1,    1],\n",
       "         [   0,   56,  309,  ...,    1,    1,    1],\n",
       "         [   0,  892,  568,  ...,    1,    1,    1],\n",
       "         ...,\n",
       "         [   0, 1543,  261,  ...,    1,    1,    1],\n",
       "         [   0, 3796,   93,  ...,    1,    1,    1],\n",
       "         [   0,  564, 2787,  ...,    1,    1,    1]])}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loader.dataset.encodings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "66d0c21c-5075-4c9b-a5af-93f366c065df",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/christianbromley/.virtualenvs/machine-learning-course/lib/python3.9/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# build RoBERTa config\n",
    "config = RobertaConfig(\n",
    "    vocab_size=30_522,  # we align this to the tokenizer vocab_size\n",
    "    max_position_embeddings=514,\n",
    "    hidden_size=768,\n",
    "    num_attention_heads=12,\n",
    "    num_hidden_layers=6,\n",
    "    type_vocab_size=1\n",
    ")\n",
    "\n",
    "\n",
    "# init model\n",
    "model = RobertaForMaskedLM(config)\n",
    "# set device\n",
    "device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')\n",
    "# and move our model over to the selected device\n",
    "model.to(device)\n",
    "# activate training mode\n",
    "model.train()\n",
    "# initialize optimizer\n",
    "optim = AdamW(model.parameters(), lr=1e-4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7a5c8b0b-75f8-41e0-a1c3-bca26442d757",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set number of epochs\n",
    "epochs = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85c4341e-2d7c-4f49-9684-887c9a55b691",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "go\n"
     ]
    }
   ],
   "source": [
    "for epoch in range(epochs):\n",
    "    input_ids = loader.dataset.encodings['input_ids'].to(device)\n",
    "    attention_mask = loader.dataset.encodings['mask'].to(device)\n",
    "    labels = loader.dataset.encodings['labels'].to(device)\n",
    "    print('go')\n",
    "    # model\n",
    "    outputs = model(input_ids,\n",
    "                    attention_mask=attention_mask,\n",
    "                    labels=labels)\n",
    "    # extract the loss\n",
    "    loss = outputs.loss\n",
    "    # calculate loss for every parameter that needs grad update\n",
    "    loss.backward()\n",
    "    # update parameters\n",
    "    optim.step()\n",
    "    # print relevant info to progress bar\n",
    "    loop.set_description(f'Epoch {epoch}')\n",
    "    loop.set_postfix(loss=loss.item())\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "0147638d-aa84-4f59-bbce-209fc8952e64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0it [00:00, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "116\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'int' object has no attribute 'keys'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/dy/gj6r2mpj09b9nq4f9qll_b000000gp/T/ipykernel_18444/3453229787.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      8\u001b[0m         \u001b[0moptim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mzero_grad\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m         \u001b[0;31m# pull all tensor batches required for training\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbatch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m         \u001b[0minput_ids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbatch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'input_ids'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0mattention_mask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbatch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'mask'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'int' object has no attribute 'keys'"
     ]
    }
   ],
   "source": [
    "#loop = tqdm(loader, leave=True)\n",
    "# loop through epochs\n",
    "for epoch in range(epochs):\n",
    "    print(f'Epoch {str(epoch)}')\n",
    "    for batch, feed_dict in tqdm.tqdm(enumerate(loader)):\n",
    "        print(i_batch)\n",
    "        # initialize calculated gradients (from prev step)\n",
    "        optim.zero_grad()\n",
    "        # pull all tensor batches required for training\n",
    "        #print(batch.keys())\n",
    "        input_ids = batch['input_ids'][0].to(device)\n",
    "        attention_mask = batch['mask'][0].to(device)\n",
    "        labels = batch['labels'][0].to(device)\n",
    "        # model\n",
    "        outputs = model(input_ids,\n",
    "                        attention_mask=attention_mask,\n",
    "                        labels=labels)\n",
    "        # extract the loss\n",
    "        loss = outputs.loss\n",
    "        # calculate loss for every parameter that needs grad update\n",
    "        loss.backward()\n",
    "        # update parameters\n",
    "        optim.step()\n",
    "        # print relevant info to progress bar\n",
    "        loop.set_description(f'Epoch {epoch}')\n",
    "        loop.set_postfix(loss=loss.item())\n",
    "        \n",
    "model.save_pretrained(f'{output_dir}/thesis_bert')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f1801130-5fe9-41d0-8f52-58bd3a4d0dac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 0\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "'module' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/dy/gj6r2mpj09b9nq4f9qll_b000000gp/T/ipykernel_18444/3104237525.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'Epoch {str(epoch)}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;31m# setup loop with TQDM and dataloader\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m     \u001b[0mloop\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtqdm\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloader\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mleave\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloop\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mbatch\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mloop\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'module' object is not callable"
     ]
    }
   ],
   "source": [
    "# loop through epochs\n",
    "for epoch in range(epochs):\n",
    "    print(f'Epoch {str(epoch)}')\n",
    "    # setup loop with TQDM and dataloader\n",
    "    loop = tqdm(loader, leave=True)\n",
    "    print(loop)\n",
    "    for batch in loop:\n",
    "        print(batch.keys())\n",
    "        # initialize calculated gradients (from prev step)\n",
    "        optim.zero_grad()\n",
    "        # pull all tensor batches required for training\n",
    "        print(batch.keys())\n",
    "        input_ids = batch['input_ids'][0].to(device)\n",
    "        attention_mask = batch['mask'][0].to(device)\n",
    "        labels = batch['labels'][0].to(device)\n",
    "        # model\n",
    "        outputs = model(input_ids,\n",
    "                        attention_mask=attention_mask,\n",
    "                        labels=labels)\n",
    "        # extract the loss\n",
    "        loss = outputs.loss\n",
    "        # calculate loss for every parameter that needs grad update\n",
    "        loss.backward()\n",
    "        # update parameters\n",
    "        optim.step()\n",
    "        # print relevant info to progress bar\n",
    "        loop.set_description(f'Epoch {epoch}')\n",
    "        loop.set_postfix(loss=loss.item())\n",
    "        \n",
    "model.save_pretrained(f'{output_dir}/thesis_bert')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "40ed3f66-fd01-44d1-84f2-85534d416e4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_model(config, loader):\n",
    "    # init model\n",
    "    model = RobertaForMaskedLM(config)\n",
    "    # set device\n",
    "    device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')\n",
    "    # and move our model over to the selected device\n",
    "    model.to(device)\n",
    "    # activate training mode\n",
    "    model.train()\n",
    "    # initialize optimizer\n",
    "    optim = AdamW(model.parameters(), lr=1e-4)\n",
    "    # set number of epochs\n",
    "    epochs = 2\n",
    "    # loop through epochs\n",
    "    for epoch in range(epochs):\n",
    "        print(f'Epoch {str(epoch)}')\n",
    "        # setup loop with TQDM and dataloader\n",
    "        loop = tqdm(loader, leave=True)\n",
    "        for batch in loop:\n",
    "            print(batch.keys())\n",
    "            # initialize calculated gradients (from prev step)\n",
    "            optim.zero_grad()\n",
    "            # pull all tensor batches required for training\n",
    "            print(batch.keys())\n",
    "            input_ids = batch['input_ids'][0].to(device)\n",
    "            attention_mask = batch['mask'][0].to(device)\n",
    "            labels = batch['labels'][0].to(device)\n",
    "            # model\n",
    "            outputs = model(input_ids,\n",
    "                            attention_mask=attention_mask,\n",
    "                            labels=labels)\n",
    "            # extract the loss\n",
    "            loss = outputs.loss\n",
    "            # calculate loss for every parameter that needs grad update\n",
    "            loss.backward()\n",
    "            # update parameters\n",
    "            optim.step()\n",
    "            # print relevant info to progress bar\n",
    "            loop.set_description(f'Epoch {epoch}')\n",
    "            loop.set_postfix(loss=loss.item())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "c16c966b-f2ae-4336-95ca-04328528b70e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                                                                                                                                           | 0/117 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['input_ids'])\n",
      "dict_keys(['input_ids'])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'mask'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/dy/gj6r2mpj09b9nq4f9qll_b000000gp/T/ipykernel_18444/868245220.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m )\n\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m \u001b[0mtrain_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloader\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave_pretrained\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'{output_dir}/thesis_bert'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/dy/gj6r2mpj09b9nq4f9qll_b000000gp/T/ipykernel_18444/3398301872.py\u001b[0m in \u001b[0;36mtrain_model\u001b[0;34m(config, loader)\u001b[0m\n\u001b[1;32m     24\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbatch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m             \u001b[0minput_ids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbatch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'input_ids'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m             \u001b[0mattention_mask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbatch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'mask'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m             \u001b[0mlabels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbatch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'labels'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m             \u001b[0;31m# model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'mask'"
     ]
    }
   ],
   "source": [
    "\n",
    "train_model(config, loader)\n",
    "\n",
    "model.save_pretrained(f'{output_dir}/thesis_bert')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b685612-7889-42c5-9458-2d7ebd24f6ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
